---
title: "Benchmark enrichment MassExpression"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Functions

```{r}
library(ggplot2)
library(dplyr)
library(tidyr)
library(rjson)
library(EnrichmentAnalysisStepR)

get_the_gmt_folder <- function(enrichment_settings_file, gene_sets_folder) {

  species_settings <- fromJSON(enrichment_settings_file)
  species <- as.matrix(species_settings$species)


  taxonomy_id <- switch(
    species,
    "Human"= "9606",
    "Mouse"= "10090",
    "Yeast"= "559292",
    "Unknown"
  )

  if (taxonomy_id == "Unknown") {
    return("Unknown")
  }

  paste(gene_sets_folder, taxonomy_id, sep="")
}

assign_uniprot_ids_mq <- function(proteinGroups){
  
  uniprot_pattern = "[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}"
  major_ids <- str_extract_all(proteinGroups$Majority.protein.IDs, uniprot_pattern)
  num_major_ids <- table(unlist(lapply(major_ids, length)))
  selected_ids <- unlist(lapply(major_ids, function(x){x[1]}))
  
  gene_patterns = str_c(selected_ids, ".*GN=(\\S*) ")
  string_id_to_gene = str_extract(proteinGroups$Fasta.headers, gene_patterns) # get the string for the right assession
  string_just_gene = str_extract(string_id_to_gene, "GN=(\\S*) ")
  genes = gsub("[GN=| ]","", string_just_gene)
  
  string_id_to_gene = str_extract(proteinGroups$Fasta.headers, gene_patterns) # get the string for the right assession
  string_description = str_extract(string_id_to_gene, " (.*) ")
  descriptions = gsub("GN.*","", string_description )
  
  
  proteinGroups$ProteinId <- selected_ids
  proteinGroups$GeneId <- genes
  proteinGroups$Description <- descriptions
  
  return(proteinGroups)
}


make_long_wide_df <- function(matrix_prot, new_int_col = "IntME"){
  int_prot <- data.frame(matrix_prot)
  int_prot$ProteinId <- rownames(int_prot)
  long_int_me <- int_prot %>% pivot_longer(cols = c(LFQ.intensity.1_hu_C1:LFQ.intensity.6_hu_P3), 
                                      values_to = new_int_col, names_to = "SampleName")
  long_int_me$SampleName <- gsub("LFQ.intensity.","", long_int_me$SampleName)
  long_int_me$SampleName <- tolower(long_int_me$SampleName)
  return(long_int_me)
}

make_long_wide_df_mq <- function(matrix_prot, new_int_col = "IntMQ"){
  int_prot <- data.frame(matrix_prot)
  int_prot$ProteinId <- rownames(int_prot)
  long_int_me <- int_prot %>% pivot_longer(cols = c(Parent.1, Parent.2, Parent.3, Control.1, Control.2 , Control.3), values_to = new_int_col, names_to = "run_id")
  return(long_int_me)
}
```

## Input folders

```{r generic-import}
# bench-enrichment folder is on Anna's laptop, not pushed on github
# data downloaded from S3 containing the data produced by running HER2 with the maxquant upload
gene_sets_folder <- "../../enrichment-step/gene_sets/"
enrich_setting_mq <- "../../universal-import/MassExpression/dev/bench-enrichment/upload-maxquant/enrichment_settings.json"
enrich_setting_me <- "../../universal-import/MassExpression/dev/bench-enrichment/upload/enrichment_settings.json"

# data downloaded from S3 containing the data produced by running HER2 with the maxquant upload
# upload_folder_me <- "../../universal-import/MassExpression/dev/bench-enrichment/upload"

# DiscoveryData run locally after updating MassExpression
upload_folder_me <- "../../universal-import/MassExpression/dev/"
upload_folder_mq <- "../../universal-import/MassExpression/dev/bench-enrichment/transform-maxquant/"

out_folder_me <- "../../universal-import/MassExpression/dev/bench-enrichment/out_me"
out_folder_mq <- "../../universal-import/MassExpression/dev/bench-enrichment/out_mq"
```


```{r}
get_the_gmt_folder(enrich_setting_mq, gene_sets_folder)
```

```{r}
get_the_gmt_folder(enrichment_settings_file = enrich_setting_me,  gene_sets_folder = gene_sets_folder)
```

```{r main-r}
# install.packages("rjson", repos="http://cran.rstudio.com/")

#name pars
input_folder <- upload_folder_me
output_folder <- out_folder_me
enrichment_settings_file <- enrich_setting_me

gmt_folder <- get_the_gmt_folder(enrich_setting_me, gene_sets_folder)

if(gmt_folder != "Unknown") {
  results = runEnrichmentWorkflowStep(input_folder,
                                     gmtFolder = gmt_folder,
                                     outputFolder = output_folder,
                                     by = "Protein",
                                     method = "camera")
}

results_me <- results
```

```{r}
#name pars
input_folder <- upload_folder_mq
output_folder <- out_folder_mq
enrichment_settings_file <- enrich_setting_mq

gmt_folder <- get_the_gmt_folder(enrichment_settings_file, gene_sets_folder)

if(gmt_folder != "Unknown") {
  results_mq_1 = runEnrichmentWorkflowStep(input_folder,
                                     gmtFolder = gmt_folder,
                                     outputFolder = output_folder,
                                     by = "Protein",
                                     method = "camera")
}

run_second_time = FALSE
if(run_second_time){
  results_mq_2 = runEnrichmentWorkflowStep(input_folder,
                                     gmtFolder = gmt_folder,
                                     outputFolder = output_folder,
                                     by = "Protein",
                                     method = "camera")
}
```

## Reproduce error

```{r}
bp_me <- results_me[[1]]$libraryStatistics[[1]]
bp_mq_1 <- results_mq_1[[1]]$libraryStatistics[[1]]
# bp_mq_2 <- results_mq_2[[1]]$libraryStatistics[[2]]

bp_me <- bp_me %>% dplyr::select(gene.set, pval, afc, observed, size) %>% 
  dplyr::rename(pval_me = pval, afc_me = afc, observed_me = observed, size_me = size)
bp_mq_1 <- bp_mq_1 %>% dplyr::select(gene.set, pval, afc, observed, size) %>% 
  dplyr::rename(pval_mq_1 = pval, afc_mq_1 = afc, observed_mq_1 = observed, size_mq_1 = size)
#bp_mq_2 <- bp_mq_2 %>% dplyr::select(gene.set, pval, afc, observed, size) %>% 
 # dplyr::rename(pval_mq_2 = pval, afc_mq_2 = afc, observed_mq_2 = observed, size_mq_2 = size)
```

```{r}
sum(!(bp_mq_1$gene.set %in% bp_me$gene.set))
sum(!(bp_me$gene.set %in% bp_mq_1$gene.set))

gene_set_compare <- bp_mq_1 %>% left_join(bp_me)

#gene_set_compare_two_mq <- bp_mq_1 %>% left_join(bp_mq_2)
```

```{r}
create_df_compare_enrichment <- function(results_me, results_mq_1, index){
  bp_me <- results_me[[1]]$libraryStatistics[[index]]
  bp_mq_1 <- results_mq_1[[1]]$libraryStatistics[[index]]
  bp_me <- bp_me %>% dplyr::select(gene.set, pval, afc, observed, size) %>% 
    dplyr::rename(pval_me = pval, afc_me = afc, observed_me = observed, size_me = size)
  bp_mq_1 <- bp_mq_1 %>% dplyr::select(gene.set, pval, afc, observed, size) %>% 
    dplyr::rename(pval_mq_1 = pval, afc_mq_1 = afc, observed_mq_1 = observed, size_mq_1 = size)
  list(bp_me=bp_me, bp_mq_1=bp_mq_1)
}
```


```{r}
out <- create_df_compare_enrichment(results_me, results_mq_1, "BiologicalProcess")
bp_mq_1 <- out$bp_mq_1
bp_me <- out$bp_me
compare_en <- bp_mq_1 %>% left_join(bp_me)
title <- "Biological processes"

par(mfrow=c(1,2))
plot(compare_en$afc_mq, compare_en$afc_me, main=paste0("logFC\n", title))
plot(-log10(compare_en$pval_mq), -log10(compare_en$pval_me), main="pval")
```


```{r}
out <- create_df_compare_enrichment(results_me, results_mq_1, "CellularComponents")
bp_mq_1 <- out$bp_mq_1
bp_me <- out$bp_me
title <- "CellularComponents"

compare_en <- bp_mq_1 %>% left_join(bp_me)
par(mfrow=c(1,2))
plot(compare_en$afc_mq, compare_en$afc_me, main=paste0("logFC\n", title))
plot(-log10(compare_en$pval_mq), -log10(compare_en$pval_me), main="pval")
```


```{r}
out <- create_df_compare_enrichment(results_me, results_mq_1, "MolecularFunction")
bp_mq_1 <- out$bp_mq_1
bp_me <- out$bp_me
title <- "MolecularFunction"
compare_en <- bp_mq_1 %>% left_join(bp_me)
par(mfrow=c(1,2))
plot(compare_en$afc_mq, compare_en$afc_me, main=paste0("logFC\n", title))
plot(-log10(compare_en$pval_mq), -log10(compare_en$pval_me), main="pval")
```


```{r}
out <- create_df_compare_enrichment(results_me, results_mq_1, "Reactome")
bp_mq_1 <- out$bp_mq_1
bp_me <- out$bp_me
title <- "Reactome"
compare_en <- bp_mq_1 %>% left_join(bp_me)
par(mfrow=c(1,2))
plot(compare_en$afc_mq, compare_en$afc_me, main=paste0("logFC\n", title))
plot(-log10(compare_en$pval_mq), -log10(compare_en$pval_me), main="pval")
```


```{r}
par(mfrow=c(2,1))
plot(gene_set_compare$afc_mq, gene_set_compare$afc_me, ylab="AFC-MassExpr", xlab="AFC-MQ")
abline(a = 0, b=1)

plot(-log10(gene_set_compare$pval_mq), -log10(gene_set_compare$pval_me),
     ylab="-log10p-MassExpr", xlab="-log10p-MQ")
abline(a = 0, b=1)
```


```{r}
par(mfrow=c(1,2))
plot(gene_set_compare_two_mq$afc_mq_1, gene_set_compare_two_mq$afc_mq_2, main="Two MQ runs")
abline(a = 0, b=1)

plot(gene_set_compare_two_mq$pval_mq_1, gene_set_compare_two_mq$pval_mq_2)
abline(a = 0, b=1)
```


```{r}
plot(gene_set_compare$observed_mq_1, gene_set_compare$observed_me, main="gene set sizes")
abline(a = 0, b=1)
```

### One finding

- the AFC are basically the same but the pvalues are different

## Compare input intensities mass expression

```{r}
load(file.path("../../universal-import/MassExpression/dev/DiscoveryQuant.RData"))
complete_app <- CompleteIntensityExperiment
complete_app_comp <- comparisonExperiments[[1]]

complete_app_long <- make_long_wide_df(data.frame(assay(complete_app)))
complete_app_long_comp <- make_long_wide_df(data.frame(assay(complete_app_comp)), new_int_col = "IntMEComp")

compare_me <- complete_app_long_comp %>% left_join(complete_app_long)

ggplot(compare_me, aes(x = IntME, IntMEComp)) + geom_point()

compare_me[compare_me$ProteinId %in% "P07858",]

assay(complete_app_comp)[rownames(complete_app_comp) %in% "P07858",]
```


## Compare input intensities ME and MaxQuant

```{r}
# MQ
match_id_mq <- data.frame(readRDS("match_id.rds")) %>% dplyr::rename(SampleName = mqExperiment)

mdQuantExports <- loadMDQuantOutput(upload_folder_mq)
comparisonExperiments <- data.frame(assay(listcomparisonExperimentsList(mdQuantExports)[[1]]))
comp_mq <- make_long_wide_df_mq(comparisonExperiments)

# int_impute = log2Int
prot_long_mq <- as_tibble(comp_mq) %>%
  left_join(match_id_mq) 

compare_me_total <- prot_long_mq %>% left_join(complete_app_long)
ggplot(compare_me_total, aes(x = IntME, IntMQ)) + geom_point()

compare_me_total[compare_me_total$ProteinId %in% "P07858",]
```

```{r}
compare_me_comp <- prot_long_mq %>% left_join(complete_app_long_comp)
ggplot(compare_me_comp, aes(x = IntMEComp, IntMQ)) + geom_point()
```


```{r}
compare_me_comp$different <- (compare_me_comp$IntME - compare$IntMQ) > 10e-5
saveRDS(compare$ProteinId[compare$different], "protein_different_prob_imputed.rds")
```


Input data is ok!

## Run camera

```{r}
data_mq <- load("camera_data_mq.RData")
data_me <- load("camera_data_me.RData")
match_id_mq <- data.frame(readRDS("match_id.rds")) %>% dplyr::rename(SampleName = mqExperiment)
load("../../universal-import/MassExpression/dev/DiscoveryQuant.RData")
diff_proteins <- readRDS("protein_different_prob_imputed.rds")
```

### Expression data used for camera

```{r}
# MQ
int_mq <- data.frame(save_list_mq$expression_data)
comp_mq <- make_long_wide_df_mq(int_mq)
comp_mq <- comp_mq %>% left_join(match_id_mq)

# ME
# sone prot have been filtered because not in gene sets
long_int_me <- make_long_wide_df(save_list$expression_data)

compare_me_mq <- comp_mq %>% left_join(long_int_me)
```

The data that comes from the generic importer that is used for the camera enrichment is the same as the data produced by mass expression. 

```{r}
compare_me_mq$imputed_protein <- compare_me_mq$ProteinId %in% diff_proteins
p=ggplot(compare_me_mq, aes(x=IntMQ, y=IntME, label=ProteinId, colour=imputed_protein)) + geom_point(alpha=0.5)
ggplotly(p)
# A0FGR8
# A6NDG6
# P07858
compare_me_mq[compare_me_mq$ProteinId %in% "P07858",]
```


## Compare gene sets BiologicalProcess_set_info

```{r}
filtered_gs_me <- readRDS("filtered_geneSets_me.rds")
filtered_gs_mq <- readRDS("filtered_geneSets_mq.rds")

compare_sets <- NULL
for(setn in 1:length(filtered_gs_mq)){
  if(length(filtered_gs_me[[setn]]) ==0){
    min_equal <- as.numeric(length(filtered_gs_me[[setn]])==length(filtered_gs_mq[[setn]]))
  }else{
    min_equal <- min(filtered_gs_me[[setn]]==filtered_gs_mq[[setn]])
  }
  compare_sets <- c(compare_sets, min_equal)
}
```

- Only 9 out of > 12k sets are different
> table(compare_sets)
compare_sets
    0     1 
    9 12392 
    
    
## Compare intensities loaded

```{r}
expr_me <- readRDS("comparisonExperiment_enrichComparison_SE.rds")
expr_me <- data.frame(assay(expr_me))
expr_mq <- data.frame(readRDS("expression_data_mq_first_camera.rds"))
match_id_mq <- data.frame(readRDS("match_id.rds")) %>% dplyr::rename(SampleName = mqExperiment)

sum(!(rownames(expr_me) %in% rownames(expr_mq)))
sum(!(rownames(expr_mq) %in% rownames(expr_me)))

rownames(expr_mq)[!(rownames(expr_mq) %in% rownames(expr_me))]
# Q9UID3 - pretty low counts

rownames(expr_me)[!(rownames(expr_me) %in% rownames(expr_mq))]
```


```{r}
expr_mq$ProteinId <- rownames(expr_mq)
expr_me$ProteinId <- rownames(expr_me)

long_mq <- expr_mq %>% pivot_longer(cols = Parent.1:Control.3, 
                                    values_to = "IntMQ", names_to = "run_id")
long_mq <- long_mq %>% left_join(match_id_mq)

long_me <- expr_me %>% pivot_longer(cols = LFQ.intensity.1_hu_C1:LFQ.intensity.6_hu_P3,
                                     values_to = "IntME", names_to = "SampleName")
long_me$SampleName <- gsub("LFQ.intensity.","", long_me$SampleName)
long_me$SampleName <- tolower(long_me$SampleName)


long_compare <- long_mq %>% left_join(long_me)

ggplot(long_compare, aes(x = IntMQ, IntME)) + geom_point()
```
    
## Comapre expression data used in biological processes gene set testing

```{r}
expr_me <- data.frame(readRDS("expression_data_me_first_camera.rds"))
expr_mq <- data.frame(readRDS("expression_data_mq_first_camera.rds"))
match_id_mq <- data.frame(readRDS("match_id.rds")) %>% dplyr::rename(SampleName = mqExperiment)

sum(!(rownames(expr_me) %in% rownames(expr_mq)))
sum(!(rownames(expr_mq) %in% rownames(expr_me)))

rownames(expr_mq)[!(rownames(expr_mq) %in% rownames(expr_me))]
# Q9UID3 - pretty low counts
```


```{r}
expr_mq$ProteinId <- rownames(expr_mq)
expr_me$ProteinId <- rownames(expr_me)

long_mq <- expr_mq %>% pivot_longer(cols = Parent.1:Control.3, 
                                    values_to = "IntMQ", names_to = "run_id")
long_mq <- long_mq %>% left_join(match_id_mq)

long_me <- expr_me %>% pivot_longer(cols = LFQ.intensity.1_hu_C1:LFQ.intensity.6_hu_P3,
                                     values_to = "IntME", names_to = "SampleName")
long_me$SampleName <- gsub("LFQ.intensity.","", long_me$SampleName)
long_me$SampleName <- tolower(long_me$SampleName)


long_compare <- long_mq %>% left_join(long_me)

ggplot(long_compare, aes(x = IntMQ, IntME)) + geom_point()
```

## Compare input intensities with pre and after ME rows filtering

```{r}
# Pre row filter ME
pre_row_filter_me <- readRDS("../../universal-import/MassExpression/dev/comparisonIntensityExperiment_pre_filter_rows_ME.rds")
pre_row_filter_me <- data.frame(assay(pre_row_filter_me))
pre_row_filter_me$ProteinId <- rownames(pre_row_filter_me)
pre_long_me <- pre_row_filter_me %>% pivot_longer(cols = LFQ.intensity.1_hu_C1:LFQ.intensity.6_hu_P3,
                                     values_to = "BeforeIntME", names_to = "SampleName")
pre_long_me$SampleName <- gsub("LFQ.intensity.","", pre_long_me$SampleName)
pre_long_me$SampleName <- tolower(pre_long_me$SampleName)

# After row filter ME
after_row_filter_me <- readRDS("../../universal-import/MassExpression/dev/comparisonIntensityExperiment_after_filter_rows_ME.rds")
after_row_filter_me <- data.frame(assay(after_row_filter_me))
after_row_filter_me$ProteinId <- rownames(after_row_filter_me)
after_long_me <- after_row_filter_me %>% pivot_longer(cols = LFQ.intensity.1_hu_C1:LFQ.intensity.6_hu_P3,
                                     values_to = "AfterIntME", names_to = "SampleName")
after_long_me$SampleName <- gsub("LFQ.intensity.","", after_long_me$SampleName)
after_long_me$SampleName <- tolower(after_long_me$SampleName)
```

```{r within-me}
compare_pre_after_me <- pre_long_me %>% left_join(after_long_me)

plot(compare_pre_after_me$BeforeIntME, compare_pre_after_me$AfterIntME)
```


```{r}
protIntLong <- readRDS("proteinIntenLong_parser_md_MQ.rds")

mq_with_ids <- readRDS("../../universal-import/check_imputation_steps/merge_dt.rds") %>%
  dplyr::select(`majority protein ids`, `fasta headers`, id, mqExperiment)

# int_impute = log2Int
protIntLong <- as_tibble(protIntLong) %>%
  left_join(mq_with_ids) %>%
  dplyr::rename(Majority.protein.IDs=`majority protein ids`,
                Fasta.headers=`fasta headers`) %>%
  dplyr::rename(SampleName = mqExperiment)

mq_1prot <- assign_uniprot_ids_mq(protIntLong)
mq_1prot <- mq_1prot %>%  
  dplyr::select("ProteinId", "SampleName", "log2NInt", "Imputed") 

# merge
mq_1prot <- mq_1prot %>% left_join(compare_pre_after_me)

ggplot(mq_1prot[!is.na(mq_1prot$AfterIntME), ], aes(x=log2NInt, y = AfterIntME)) + geom_point()
```



