---
title: "AQ-test-experiment"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data 

```{r}
library(EnrichmentAnalysisStepR)
```

```{r}
library(ggplot2)
library(dplyr)
library(tidyr)
library(rjson)
library(EnrichmentAnalysisStepR)

get_the_gmt_folder <- function(enrichment_settings_file, gene_sets_folder) {

  species_settings <- fromJSON(enrichment_settings_file)
  species <- as.matrix(species_settings$species)


  taxonomy_id <- switch(
    species,
    "Human"= "9606",
    "Mouse"= "10090",
    "Yeast"= "559292",
    "Unknown"
  )

  if (taxonomy_id == "Unknown") {
    return("Unknown")
  }

  paste(gene_sets_folder, taxonomy_id, sep="")
}

assign_uniprot_ids_mq <- function(proteinGroups){
  
  uniprot_pattern = "[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}"
  major_ids <- str_extract_all(proteinGroups$Majority.protein.IDs, uniprot_pattern)
  num_major_ids <- table(unlist(lapply(major_ids, length)))
  selected_ids <- unlist(lapply(major_ids, function(x){x[1]}))
  
  gene_patterns = str_c(selected_ids, ".*GN=(\\S*) ")
  string_id_to_gene = str_extract(proteinGroups$Fasta.headers, gene_patterns) # get the string for the right assession
  string_just_gene = str_extract(string_id_to_gene, "GN=(\\S*) ")
  genes = gsub("[GN=| ]","", string_just_gene)
  
  string_id_to_gene = str_extract(proteinGroups$Fasta.headers, gene_patterns) # get the string for the right assession
  string_description = str_extract(string_id_to_gene, " (.*) ")
  descriptions = gsub("GN.*","", string_description )
  
  
  proteinGroups$ProteinId <- selected_ids
  proteinGroups$GeneId <- genes
  proteinGroups$Description <- descriptions
  
  return(proteinGroups)
}


make_long_wide_df <- function(matrix_prot, new_int_col = "IntME"){
  int_prot <- data.frame(matrix_prot)
  int_prot$ProteinId <- rownames(int_prot)
  long_int_me <- int_prot %>% pivot_longer(cols = c(LFQ.intensity.1_hu_C1:LFQ.intensity.6_hu_P3), 
                                      values_to = new_int_col, names_to = "SampleName")
  long_int_me$SampleName <- gsub("LFQ.intensity.","", long_int_me$SampleName)
  long_int_me$SampleName <- tolower(long_int_me$SampleName)
  return(long_int_me)
}

make_long_wide_df_mq <- function(matrix_prot, new_int_col = "IntMQ"){
  int_prot <- data.frame(matrix_prot)
  int_prot$ProteinId <- rownames(int_prot)
  long_int_me <- int_prot %>% pivot_longer(cols = c(Parent.1, Parent.2, Parent.3, Control.1, Control.2 , Control.3), values_to = new_int_col, names_to = "run_id")
  return(long_int_me)
}
```


```{r}
gene_sets_folder <- "../../enrichment-step/gene_sets/"
enrich_setting_mq <- "../../../user-data/joel/data/enrichment_settings.json"
transform_folder <- "../../../user-data/joel/data/"
output_folder <- "../../../user-data/joel/output/"

```

```{r main-r}
# install.packages("rjson", repos="http://cran.rstudio.com/")

#name pars
input_folder <- transform_folder
output_folder <- output_folder
enrichment_settings_file <- enrich_setting_mq

upload_folder_me <- "../../universal-import/MassExpression/dev/bench-enrichment/transform-maxquant/"

gmt_folder <- "../../enrichment-step/gene_sets/9606"

if(gmt_folder != "Unknown") {
 debugonce(runEnrichmentWorkflowStep(input_folder,
                                     gmtFolder = gmt_folder,
                                     outputFolder = output_folder,
                                     by = "Protein",
                                     method = "camera"))
}

results_me <- results
```


# Test experiment

```{bash }

```


```{r}

```

