---
title: "Enrichment Benchmarking"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## HER Datase

```{r}

library(EnrichmentAnalysisStepR)
experiment_home = "/home/joseph/experiments/her_data_gsea_test"
gmt_folder = "/home/joseph/experiments/BKG_gmt_output_human/9606"



results_ora = enrichment_workflow_step(experiment_home, 
                                   gmt_folder = gmt_folder,
                                   output_folder = file.path(experiment_home,"ora"),
                                   by = "Protein",
                                   method = "ora")

experiment_home = "/home/joseph/experiments/her_data_gsea_test"
gmt_folder = "/home/joseph/experiments/BKG_gmt_output_human/9606"
results_gsea = enrichment_workflow_step(experiment_home, 
                                   gmt_folder = gmt_folder,
                                   output_folder = file.path(experiment_home,"gsea"),
                                   by = "Protein",
                                   method = "gsea")



```

```{r}
View(results_ora$`C - P`$BiologicalProcess)
```

```{r}
results_gsea$`C - P`$BiologicalProcess
```


## Write Artificial GMT File

```{r}
upload_folder <-  "/home/joseph/experiments/ups_benchmark/"
by = "Protein"


# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)

protein_count_and_intensities = read_json(file.path(upload_folder,"protein_counts_and_intensity.json"), simplifyVector = TRUE)
protein_count_and_intensities = process_protein_intensities(protein_count_and_intensities, by = by)
stopifnot(dim(protein_count_and_intensities)[1]>0)

protein_count_and_intensities$ProteinId
```

```{r}
gmt_folder = "/home/joseph/experiments/BKG_gmt_output_yeast/559292/"
```

```{r}

upload_folder = "/home/joseph/experiments/iprg2015_enr_benchmark/"
gmt_folder = "/home/joseph/experiments/benchmark_gmts/yeast"
method = "gsea"
by = "Protein"
perm = 10

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)

protein_count_and_intensities = read_json(file.path(upload_folder,"protein_counts_and_intensity.json"), simplifyVector = TRUE)
protein_count_and_intensities = process_protein_intensities(protein_count_and_intensities, by = by)
stopifnot(dim(protein_count_and_intensities)[1]>0)

# for comparison in comparisons...

results = list()
for (row in (1:nrow(protein_viz))[1]){

  comparison = protein_viz[row,"conditionComparison"]
  print(comparison)

  comparison_viz = protein_viz[row, "data"]
  tmp_protein_int = filter_protein_ints(comparison, protein_count_and_intensities)

  cls_vec = get_cls_vec(comparison, tmp_protein_int)
  se_comparison <- protein_viz_int_2_de_exp(comparison_viz, tmp_protein_int, cls_vec, by = by)

  results[[comparison]] = perform_comparison_enrichment(se_comparison, gmt_folder, method, perm)

}
results <- enrichment_adjust_p_values(results)
results <- enrichment_annotate_results(results, gmt_folder)
#enrich_write_output_tables(results, output_folder)


# done!
results
```



```{r}
library(EnrichmentAnalysisStepR)
experiment_home = "/home/joseph/experiments/iprg2015_enr_benchmark/"
gmt_folder = "/home/joseph/experiments/benchmark_gmts/yeast"

debugonce(enrichment_adjust_p_values)
results = enrichment_workflow_step(experiment_home, 
                                   gmt_folder = gmt_folder,
                                   output_folder = file.path(experiment_home,"transform"),
                                   by = "Protein",
                                   method = "gsea",
                                   perm = 250)

```


```{r}

```


```{r No False Positives Check}

enrichment_plot <- function(enrichment_results, title){
  enrichment_results = as.data.frame(enrichment_results)
  enrichment_results$colour = 'black'
  enrichment_results$colour[enrichment_results$pval < 0.05] = 'red' 
  
  plot(enrichment_results$nes, 
     -log10(enrichment_results$pval),
     col = enrichment_results$colour,
     main = title)
}

for (i in names(results)){
  for (j in names(results[[i]])){
    enrichment_plot(results[[i]][[j]], paste(i,j))
  }
}
```

```{r}
tmp = results$`sample1 - sample2`$BiologicalProcess_iPRG2015
tmp
```


"Ovalbumin": "P44015", 
"Myoglobin": "P55752", 
"Phosphorylase_b": "P44374", 
"BetaGalactosidase": "P44983" , 
"BSA": "P44683", 
"CarbonicAnhydrase": "P55249"

Lines added to GMT:


iPRG_all	All the iprg2015 spike-ins	P44374 P44983 P51862 P55249 P55752 Q06683
iPRG_all	All the iprg2015 spike-ins	P44374 P44983 P51862 P55249 P55752 Q06683
iPRG_up_in_1_vs_2	1 vs 2 pos	P44015 P55752 P44374 P44683
iPRG_up_in_2_vs_1	2 vs 1 pos	P44983 P55249
iPRG_up_in_1_vs_2_noise	1 vs 2 noise	P44015 P55752 P44374 P44683 P32487 P40556
iPRG_up_in_2_vs_1_noise	2 vs 1 noise	P44983 P55249 P32487 P40556


All should be significant in some comparisons
1 vs 2 and 2 vs 1 should be significant in those comparisons


## Run Reactome with our data


```{r PADOG REACTOME SUBMISSION}

library(EnrichmentAnalysisStepR)


upload_folder = "/home/joseph/experiments/her_data_gsea_test"
by = "Protein"

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)

protein_count_and_intensities = read_json(file.path(upload_folder,"protein_counts_and_intensity.json"), simplifyVector = TRUE)
protein_count_and_intensities = process_protein_intensities(protein_count_and_intensities, by = "Protein")
stopifnot(dim(protein_count_and_intensities)[1]>0)

# for comparison in comparisons...

for_reactome = na.omit(protein_count_and_intensities)
write.table(for_reactome, "her_study_reactome.tsv",  sep ="\t", quote = F, row.names = F)
```

```{r}

results_padog = enrichment_workflow_step(experiment_home, 
                                   gmt_folder = gmt_folder,
                                   output_folder = file.path(experiment_home,"gsea"),
                                   by = "Protein",
                                   method = "padog")


```


# Workflow Tests

```{r Enrichment Runner, NEW}

# single condition protein
results <- enrichment_workflow_step(upload_folder = "/home/joseph/experiments/her_data_gsea_test",
                                   gmt_folder ="/home/joseph/experiments/BKG_gmt_output_human/9606",
                                   output_folder = "/home/joseph/experiments/her_data_gsea_test/padog",
                                   by = "Protein",
                                   method = "padog",
                                   perm = 25)

# multiple conditions test
experiment_home = "/home/joseph/experiments/mouse_pxd_ida"
gmt_folder = "/home/joseph/experiments/BKG_gmt_output_mouse"
results = enrichment_workflow_step(experiment_home, 
                                      gmt_folder = gmt_folder,
                                      output_folder = file.path(experiment_home,"transform"),
                                      by = "Protein",
                                   method = "padog")


# discovery test
experiment_home = "/home/joseph/experiments/ldopa_test_disco_human"
results = enrichment_workflow_step(experiment_home, 
                                      gmt_folder ="/home/joseph/experiments/BKG_gmt_output_human/9606",  
                                      output_folder = file.path(experiment_home,"transform"),
                                      by = "Protein",
                                   method = "padog")


# by gene
experiment_home = "/home/joseph/experiments/ldopa_test_disco_human"
results = enrichment_workflow_step(experiment_home, 
                                      gmt_folder ="./ckg_gmts_by_gene",  
                                      output_folder = file.path(experiment_home,"transform"),
                                      by = "Gene",
                                   method = "padog")

```

# Add Average Fold Change to output

```{r}
debugonce(perform_comparison_enrichment)
# single condition protein
results <- enrichment_workflow_step(upload_folder = "/home/joseph/experiments/her_data_gsea_test",
                                    gmt_folder ="/home/joseph/experiments/BKG_gmt_output_human/9606",
                                    output_folder = "./gsea_results",
                                    by = "Protein",
                                    method = "padog",
                                    perm = 25)

```

```{r}

results = add_gene_sets_to_result(results, gs)

# get's average fold change using a gene set and the SE artifact.
get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"])
}

add_AFC_to_results <- function(results, se_comparison){
  results$AFC <- unlist(lapply(results$items, function(x){get_AFC(se_comparison, x)}))
  results
}

results <- add_AFC_to_results(results, se_comparison)

```



```{r}
results
plot(results$AFC, -log10(results$PVAL))
```



```{r}
upload_folder = "/home/joseph/experiments/her_data_gsea_test"
gmt_folder ="/home/joseph/experiments/BKG_gmt_output_human/9606"
output_folder = "./gsea_results"
by = "Protein"
method = "padog"
perm = 25


# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# for comparison in comparisons...
results = list()
row = 1
# Get the Comparison
comparison = protein_viz[row,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[row, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)


comparison = se_comparison

enrichment = list()

for (library in list.files(gmt_folder, pattern = "\\.gmt$")){
  gmt.file <- file.path(gmt_folder, library)

  gs <- getGenesets(gmt.file)

  print(gmt.file)
  print(paste("Sets in Library: ", length(gs)))
  print(paste("Genes in Library: ", length(unique(unlist(gs)))))
  print(paste("Genes in Library and dataset: ", length(intersect(rownames(comparison),unlist(gs)))))

  # check for common ids
  if (length(intersect(rownames(comparison),unlist(gs)))>100){

    # do enrichment
    sbea.res <- sbea(method = method, se = comparison, 
                     gs = gmt.file, perm = perm, alpha = 1)
    results = as.data.frame(gsRanking(sbea.res))

    # get sets info (size/observed and merge)
    #sets_info <- add_sets_info(gs, comparison)
    #results <- merge(results,sets_info)

    # write results
    lib = str_split(library,".gmt")[[1]][1]
    label = str_c(lib,"_enrichment.csv")

    enrichment[[lib]] = add_gene_sets_to_result(results, gs)
    
  } else {
    print("Skipping enrichment, not enough genes in common.")
    print("Heres some example ids")
    print(rownames(comparison)[1:10])
    print(unlist(gs)[1:10])
  }

}


```



# reactome benchmark

What do we know?
- Our results show low, likely random overlap with theirs cf = (3,32,45,706)
- When we use their GMT file this is similar 
- Our AFC results are not 1:1. correlation about 0.85%. This either suggests there's a mistake in the groupings or that they are removing values. We know that they remove uniprot id's they don't map.  
- In the report, they quote 975 fold changes for gene/proteins however we think we upload 1046 comparisons. They may conflate some of the submitted data if it is isoform specific. They might also be removing specific isoforms. 

Ideas for what is causing differences:
- did I upload the correct data? -> check data as you upload, compare to expression object
- did they perform some kind of normalization? -> check their de results
- are my gene sets different?  -> create some summary statistics and compare gene sets
- removal of unmapped identifiers -> try removing unmapped identifiers in my analysis


## Look at our data

```{r}
submitted_data = read.csv("her_study_reactome.tsv", sep ="\t")

```

## Using proteins/ours

```{r Run algorithm}
# single condition protein
results <- enrichment_workflow_step(upload_folder = "/home/joseph/experiments/her_data_gsea_test",
                                   gmt_folder ="/home/joseph/experiments/BKG_gmt_output_human/9606",
                                   output_folder = "/home/joseph/experiments/her_data_gsea_test/padog",
                                   by = "Protein",
                                   method = "padog",
                                   perm = 25)
```

```{r Compare against first reactome test}
# load our results:
reactome_padog_results = read_json("/home/joseph/experiments/her_data_gsea_test/padog/C.P.Reactome.json",
                                   simplifyVector = T)
md_results = reactome_padog_results$gene.set.statistics


reactome_results = readxl::read_excel("reactome_results.xlsx")
#reactome_results
plot(reactome_results$av_foldchange.PROTEOMICS_INT_1, -log10(reactome_results$PValue.PROTEOMICS_INT_1))
plot(md_results$afc, -log10(md_results$pval))

both_results = merge(reactome_results, md_results, by.x = "Name", by.y = "name")

plot(both_results$av_foldchange.PROTEOMICS_INT_1, -both_results$afc)
plot(-log10(both_results$PValue.PROTEOMICS_INT_1), 
     -log10(both_results$pval))


cor(both_results$av_foldchange.PROTEOMICS_INT_1, -both_results$afc)
cor(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$pval))

```

```{r}
sum((both_results$`PValue.HER2 Resistance Study`<0.05) & (both_results$pval <0.05))
sum((both_results$`PValue.HER2 Resistance Study`<0.05) & (both_results$pval >0.05))
sum((both_results$`PValue.HER2 Resistance Study`>0.05) & (both_results$pval <0.05))
sum((both_results$`PValue.HER2 Resistance Study`>0.05) & (both_results$pval >0.05))

#both_results$Name[both_results]
```


## Using genes/their gmt

```{r Run algorithm over data}
# single condition protein
library(EnrichmentAnalysisStepR)
results <- enrichment_workflow_step(upload_folder = "/home/joseph/experiments/her_data_gsea_test",
                                   gmt_folder = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt",
                                   output_folder = "/home/joseph/experiments/her_data_gsea_test/reactome_gmt_genes",
                                   by = "Gene",
                                   method = "padog",
                                   perm = 25)
```

```{r Compare against first reactome test}
# load our results:
reactome_padog_results = read_json("/home/joseph/experiments/her_data_gsea_test/reactome_gmt_genes/C.P.ReactomePathways.json",
                                   simplifyVector = T)
md_results = reactome_padog_results$gene.set.statistics
md_results


reactome_results = readxl::read_excel("reactome_results.xlsx")
plot(reactome_results$av_foldchange.PROTEOMICS_INT_1, -log10(reactome_results$PValue.PROTEOMICS_INT_1))
plot(-md_results$afc, -log10(md_results$pval))

both_results = merge(reactome_results, md_results, by.x = "Name", by.y = "gene.set")

plot(both_results$av_foldchange.PROTEOMICS_INT_1, -both_results$afc)
plot(-log10(both_results$PValue.PROTEOMICS_INT_1), 
     -log10(both_results$pval))

cor(both_results$av_foldchange.PROTEOMICS_INT_1, -both_results$afc)
cor(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$pval))



sum((both_results$PValue.PROTEOMICS_INT_1<0.05) & (both_results$pval <0.05))
sum((both_results$PValue.PROTEOMICS_INT_1<0.05) & (both_results$pval >0.05))
sum((both_results$PValue.PROTEOMICS_INT_1>0.05) & (both_results$pval <0.05))
sum((both_results$PValue.PROTEOMICS_INT_1>0.05) & (both_results$pval >0.05))

#both_results$Name[both_results]
```

```{r Compare Against second reactome test}
# load our results:
reactome_padog_results = read_json("/home/joseph/experiments/her_data_gsea_test/reactome_gmt_genes/C.P.ReactomePathways.json",
                                   simplifyVector = T)
md_results = reactome_padog_results$gene.set.statistics
md_results

reactome_path = "/home/joseph/experiments/ReactomeHERResults/padog_results.xlsx"
reactome_results = readxl::read_excel(reactome_path)
plot(reactome_results$`av_foldchange.HER2 Resistance Study`, -log10(reactome_results$`PValue.HER2 Resistance Study`))
plot(-md_results$afc, -log10(md_results$pval))

both_results = merge(reactome_results, md_results, by.x = "Name", by.y = "gene.set")

plot(both_results$`av_foldchange.HER2 Resistance Study`, -both_results$afc)
plot(-log10(both_results$`PValue.HER2 Resistance Study`), 
     -log10(both_results$pval))

cor(both_results$`av_foldchange.HER2 Resistance Study`, -both_results$afc)
cor(-log10(both_results$`PValue.HER2 Resistance Study`), -log10(both_results$pval))


sum((both_results$`PValue.HER2 Resistance Study`<0.05) & (both_results$pval <0.05))
sum((both_results$`PValue.HER2 Resistance Study`<0.05) & (both_results$pval >0.05))
sum((both_results$`PValue.HER2 Resistance Study`>0.05) & (both_results$pval <0.05))
sum((both_results$`PValue.HER2 Resistance Study`>0.05) & (both_results$pval >0.05))
```

## Code Comparison 

This is what enrichment browser's padog call looks like. 

```{r}

# this is what
.padog <- function(se, gs, perm=1000)
{
    padog <- NULL
    isAvailable("PADOG", type="software")

    grp <- se[[configEBrowser("GRP.COL")]]
    grp <- ifelse(grp == 0, "c", "d") 
  
    blk <- NULL
    BLK.COL <- configEBrowser("BLK.COL")
    if(BLK.COL %in% colnames(colData(se))) 
        blk <- make.names(colData(se)[,BLK.COL]) 
    paired <- !is.null(blk)
  
    nmin <- configEBrowser("GS.MIN.SIZE")
    perm <- as.numeric(perm) 
 
    res <- padog(assay(se), group=grp, 
        paired=paired, block=blk, gslist=gs, Nmin=nmin, NI=perm)
  
    res.tbl <- res[, c("meanAbsT0", "padog0", "PmeanAbsT", "Ppadog")]
    colnames(res.tbl) <- c("MEAN.ABS.T0", 
        "PADOG0", "P.MEAN.ABS.T",  configEBrowser("PVAL.COL"))
    rownames(res.tbl) <- as.vector(res[,"ID"]) 
    return(res.tbl)
}

```

This is what Reactome's padog call looks like :


```{r}
process <- function(expression.data, sample.data, design, gene.indices, data.type, analysis.group.1, analysis.group.2) {
    # prepare the data
    expression.data <- prepareData(expression.data, sample.data, design, data.type)

    # padog requires the samples in a "control" and a "disease" group
    padog_group <- rep(NA, ncol(expression.data))
    padog_group[design[, analysis.group.1] == 1] <- "c"
    padog_group[design[, analysis.group.2] == 1] <- "d"

    # remove all samples that are not in the treatment groups
    samples_to_keep <- padog_group %in% c("c", "d")
    expression.data <- expression.data[, samples_to_keep]
    padog_group <- padog_group[samples_to_keep]
    sample.data <- sample.data[samples_to_keep, ]

    # convert the gene.indices back to the identifiers
    gene_identifier_set <- lapply(gene.indices, function(gene_ids) {
        rownames(expression.data)[gene_ids]
    })

    # get the sample group if set
    is_paired <- FALSE
    sample_block <- NULL

    if (nchar(sample.groups) > 0) {
        if (!sample.groups %in% colnames(sample.data)) {
            stop("Error: Failed to find defined sample.groups '", sample.groups, "' in the sample metadata. ",
                 "In the ReactomeGSA R package, this must also be specified as an 'additional_factor'")
        }

        is_paired <- TRUE
        sample_block <- sample.data[, sample.groups]
    }

    # check padog's requirements and create nice error messages
    found_genes <- length(unique(as.numeric(unlist(gene.indices))))
    if (found_genes <= 10) {
        stop("Error: PADOG requires more than 10 proteins/genes to be found in the gene sets. Only ", found_genes,
        " identifiers be mapped to Reactome's pathways")
    }

    # there must be at least 6 samples
    if (any(table(padog_group) < 3)) {
        stop("Error: PADOG requires at least 3 samples per group.")
    }

    # run PADOG
    padog_result <- padog(
        esetm = as.matrix(expression.data$E),
        group = padog_group,
        paired = is_paired,
        block = sample_block,
        gslist = gene_identifier_set,
        NI = 1000, # number of iterations
        plots = FALSE,
        Nmin = 0) # minimum gene set size

    # convert to the required result output
    colnames(padog_result) <- plyr::mapvalues(
        from = c("ID", "Size", "meanAbsT0", "padog0", "PmeanAbsT", "Ppadog"),
        to = c("Pathway", "NGenes", "MeanAbsT0", "MeanWeightT0", "PValue", "FDR"),
        x = colnames(padog_result))

    return(padog_result[, c("Pathway", "FDR", "PValue", "NGenes", "MeanAbsT0", "MeanWeightT0")])
}
```


## Let's compare their gene sets to our gene sets

```{r}
# 1 example 
example_1 = both_results[both_results$Name == "Antiviral mechanism by IFN-stimulated genes",]
example_1$NGenes.PROTEOMICS_INT_1
length(example_1$items[[1]]) 

#This seems to suggest the gene sets we are working with are very different


example_1$av_foldchange.PROTEOMICS_INT_1
example_1$afc

example_1$PValue.PROTEOMICS_INT_1
example_1$pval


```
```{r Which genes/proteins are in that pathway?}

# MD Results says
md_results[md_results$gene.set == "Antiviral mechanism by IFN-stimulated genes",]$items

# GS itself says...
gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
gs =  getGenesets(gmt.file)
gs$`Antiviral mechanism by IFN-stimulated genes`
```

MD and Reactome are in agreement... so why does padog say 24 only? Maybe they mean detections. Let's check how many we have signal for

```{r}
protein_viz = read_json("/home/joseph/experiments/her_data_gsea_test/protein_viz.json", simplifyVector = T)
protein_viz = protein_viz$data[[1]]
length(intersect(protein_viz$GeneName,gs$`Antiviral mechanism by IFN-stimulated genes`))
protein_viz = protein_viz[complete.cases(protein_viz),]
length(intersect(protein_viz$GeneName,gs$`Antiviral mechanism by IFN-stimulated genes`))

protein_viz[protein_viz$GeneName %in% intersect(protein_viz$GeneName,gs$`Antiviral mechanism by IFN-stimulated genes`), ]
```

```{r}

m <- submitted_data[,-1]
boxplot(x = as.list(as.data.frame(m)))
colMeans(m)
```

### What about removing unmapped proteins?

```{r read }
unmapped_proteins = read.csv("/home/joseph/experiments/ReactomeHERResults/not_found.csv", stringsAsFactors = F)
unmapped_proteins$Not.found
```


```{r Manual Enrichment Step HER}

upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Gene"
method = "padog"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)



#filter assay_data to remove not mapped proteins
unmapped_proteins = read.csv("/home/joseph/experiments/ReactomeHERResults/not_found.csv", stringsAsFactors = F)
unmapped_proteins$Not.found
unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
assay(se_comparison) = assay(se_comparison)[unmapped_filter,]

# perform normalization
normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
assay(se_comparison) = normalized_data


# do enrichment  
gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
gs <- getGenesets(gmt.file)

# do enrichment
sbea.res <- sbea(method = method, se = se_comparison, 
                 gs = gmt.file, perm = perm, alpha = 1)
results = as.data.frame(gsRanking(sbea.res))

# get sets info (size/observed and merge)
#sets_info <- add_sets_info(gs, comparison)
#results <- merge(results,sets_info)


results <- add_gene_sets_to_result(results, gs)
results <- add_AFC_to_results(results,se_comparison)
md_results <- results
colnames(md_results) <- tolower(colnames(md_results))      
```


```{r}
reactome_path = "/home/joseph/experiments/ReactomeHERResults/padog_results.xlsx"
reactome_results = readxl::read_excel(reactome_path)

plot(reactome_results$`av_foldchange.HER2 Resistance Study`, -log10(reactome_results$`PValue.HER2 Resistance Study`))
plot(md_results$afc, -log10(md_results$pval))

both_results = merge(reactome_results, md_results, by.x = "Name", by.y = "gene.set")

plot(both_results$`av_foldchange.HER2 Resistance Study`, -both_results$afc)
plot(-log10(both_results$`PValue.HER2 Resistance Study`), 
     -log10(both_results$pval))

cor(both_results$`av_foldchange.HER2 Resistance Study`, -both_results$afc)
cor(-log10(both_results$`PValue.HER2 Resistance Study`), -log10(both_results$pval))


sum((both_results$`PValue.HER2 Resistance Study`<0.05) & (both_results$pval <0.05))
sum((both_results$`PValue.HER2 Resistance Study`<0.05) & (both_results$pval >0.05))
sum((both_results$`PValue.HER2 Resistance Study`>0.05) & (both_results$pval <0.05))
sum((both_results$`PValue.HER2 Resistance Study`>0.05) & (both_results$pval >0.05))

```

```{r manual padog function}

expression.data = assay(se_comparison)

# get gene sets
gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
#gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP
padog_group <- rep(NA, ncol(expression.data))
padog_group[se_comparison$GROUP == 0] <- "c"
padog_group[se_comparison$GROUP == 1] <- "d"
padog_group


# restrict gene sets and summarized experiment to interacting genes
# restrict se and gs to intersecting genes
igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                            "gene sets (gs) have no gene IDs in common")
expression.data <- expression.data[igenes,]
gs <- lapply(gs, function(s) s[s %in% igenes])



# run PADOG
padog_result <- padog(
    esetm = expression.data,
    group = padog_group,
    paired = FALSE,
    block = NULL,
    gslist = gs,
    NI = 1000, # number of iterations
    plots = FALSE,
    Nmin = 0) # minimum gene set size

# convert to the required result output
colnames(padog_result) <- plyr::mapvalues(
    from = c("ID", "Size", "meanAbsT0", "padog0", "PmeanAbsT", "Ppadog"),
    to = c("Pathway", "NGenes", "MeanAbsT0", "MeanWeightT0", "PValue", "FDR"),
    x = colnames(padog_result))


padog_result

padog_result$items <- gs[match(padog_result$Name, names(gs))]


get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"])
}

se_comparison = se_comparison[igenes,]
padog_result$afc <- unlist(lapply(padog_result$items, function(x){get_AFC(se_comparison, x)}))
padog_result



# add names to replace id's if doing protein centric


```

```{r}
info_file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome_set_info.csv"
#print(paste("Using set info file: ", info_file))
info = read.csv(info_file, stringsAsFactors = F)
#colnames(info) = c("GENE.SET", "NAME", "DESCRIPTION", "ITEMS", "SIZE", "LINKOUT")

head(info)

info = info[,c(1,2)]

colnames(info) = c("Pathway","Name")
head(info)
padog_result$Pathway <- NULL

padog_result = merge(padog_result, info, by.x = "Name", by.y = "Pathway", all.x = T)
padog_result

```


```{r Benchmark Against Server Results when using padog directly/min size set to 0}
md_results = padog_result

reactome_path = "/home/joseph/experiments/ReactomeHERResults/padog_results.xlsx"
reactome_results = readxl::read_excel(reactome_path)

plot(reactome_results$`av_foldchange.HER2 Resistance Study`, -log10(reactome_results$`PValue.HER2 Resistance Study`))
plot(-md_results$afc, -log10(md_results$PValue))



both_results = merge(reactome_results, md_results)

plot(both_results$`av_foldchange.HER2 Resistance Study`, -both_results$afc)
plot(-log10(both_results$`PValue.HER2 Resistance Study`), 
     -log10(both_results$PValue))
plot(both_results$`NGenes.HER2 Resistance Study`, both_results$NGenes)
abline(0,1)

cor(both_results$`av_foldchange.HER2 Resistance Study`, -both_results$afc)
cor(-log10(both_results$`PValue.HER2 Resistance Study`), -log10(both_results$PValue))

tmp = both_results[both_results$NGenes>0,]
sum((tmp$`PValue.HER2 Resistance Study`<0.05) & (tmp$PValue <0.05))
sum((tmp$`PValue.HER2 Resistance Study`<0.05) & (tmp$PValue >0.05))
sum((tmp$`PValue.HER2 Resistance Study`>0.05) & (tmp$PValue <0.05))
sum((tmp$`PValue.HER2 Resistance Study`>0.05) & (tmp$PValue >0.05))


```


```{r}

md_sig = tmp[(tmp$FDR<0.05),]
reactome_sig = tmp[(tmp$`FDR.HER2 Resistance Study`<0.05),]
both_significant = tmp[(tmp$FDR<0.05)&(tmp$`FDR.HER2 Resistance Study`<0.05),]

print("MD")
"BAG3" %in% unlist(md_sig$items)
"YAP1" %in% unlist(md_sig$items)
"GAPDH" %in% unlist(md_sig$items)
"LGALS1" %in% unlist(md_sig$items)
"FSCN1" %in% unlist(md_sig$items)
"CLIC4" %in% unlist(md_sig$items)
"FN1" %in% unlist(md_sig$items)

print("Reactome")
"BAG3" %in% unlist(reactome_sig$items)
"YAP1" %in% unlist(reactome_sig$items)
"GAPDH" %in% unlist(reactome_sig$items)
"LGALS1" %in% unlist(reactome_sig$items)
"FSCN1" %in% unlist(reactome_sig$items)
"CLIC4" %in% unlist(reactome_sig$items)
"FN1" %in% unlist(reactome_sig$items)

print("Both")
"BAG3" %in% unlist(both_significant$items)
"YAP1" %in% unlist(both_significant$items)
"GAPDH" %in% unlist(both_significant$items)
"LGALS1" %in% unlist(both_significant$items)
"FSCN1" %in% unlist(both_significant$items)
"CLIC4" %in% unlist(both_significant$items)
"FN1" %in% unlist(both_significant$items)


```

```{r}

```




```{r Benchmark against BKG protein centric analysis using padog directly}

md_results = padog_result

reactome_path = "/home/joseph/experiments/ReactomeHERResults/padog_results.xlsx"
reactome_results = readxl::read_excel(reactome_path)

plot(reactome_results$`av_foldchange.HER2 Resistance Study`, -log10(reactome_results$`PValue.HER2 Resistance Study`))
plot(-md_results$afc, -log10(md_results$PValue))



both_results = merge(reactome_results, md_results, by.x ="Name", by.y = "Name.y")

plot(both_results$`av_foldchange.HER2 Resistance Study`, -both_results$afc)
plot(-log10(both_results$`PValue.HER2 Resistance Study`), 
     -log10(both_results$PValue))
plot(both_results$`NGenes.HER2 Resistance Study`, both_results$NGenes)
abline(0,1)

cor(both_results$`av_foldchange.HER2 Resistance Study`, -both_results$afc)
cor(-log10(both_results$`PValue.HER2 Resistance Study`), -log10(both_results$PValue))

tmp = both_results[both_results$NGenes>0,]
sum((tmp$`PValue.HER2 Resistance Study`<0.05) & (tmp$PValue <0.05))
sum((tmp$`PValue.HER2 Resistance Study`<0.05) & (tmp$PValue >0.05))
sum((tmp$`PValue.HER2 Resistance Study`>0.05) & (tmp$PValue <0.05))
sum((tmp$`PValue.HER2 Resistance Study`>0.05) & (tmp$PValue >0.05))



```

```{r helper functions}

add_gene_sets_to_result <- function(result_table, gs){
  result_table$items <- gs[match(result_table$GENE.SET, names(gs))]
  result_table
}


# get's average fold change using a gene set and the SE artifact.
#' @export get_AFC
get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"])
}

#' @export add_AFC_to_results
add_AFC_to_results <- function(results, se_comparison){
  results$AFC <- unlist(lapply(results$items, function(x){get_AFC(se_comparison, x)}))
  results
}
```



# can I get false positives easily?


```{r With genes and the reactome gmt}


upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Gene"
method = "padog"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)



#filter assay_data to remove not mapped proteins
unmapped_proteins = read.csv("/home/joseph/experiments/ReactomeHERResults/not_found.csv", stringsAsFactors = F)
unmapped_proteins$Not.found
unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
assay(se_comparison) = assay(se_comparison)[unmapped_filter,]

# perform normalization
normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
assay(se_comparison) = normalized_data

expression.data = assay(se_comparison)

# get gene sets
gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
#gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP
padog_group <- rep(NA, ncol(expression.data))
padog_group[se_comparison$GROUP == 0] <- "c"
padog_group[se_comparison$GROUP == 1] <- "d"
padog_group
print("not using real padog group, using this instead")
padog_group = sample(padog_group)
print(padog_group)
# restrict gene sets and summarized experiment to interacting genes
# restrict se and gs to intersecting genes
igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                            "gene sets (gs) have no gene IDs in common")
expression.data <- expression.data[igenes,]
gs <- lapply(gs, function(s) s[s %in% igenes])

print("randomizing rows as well")
rownames(expression.data) <- sample(rownames(expression.data))

# run PADOG
padog_result <- padog(
    esetm = dummy_data,
    group = padog_group,
    paired = FALSE,
    block = NULL,
    gslist = gs,
    NI = 1000, # number of iterations
    plots = FALSE,
    Nmin = 0) # minimum gene set size

# convert to the required result output
colnames(padog_result) <- plyr::mapvalues(
    from = c("ID", "Size", "meanAbsT0", "padog0", "PmeanAbsT", "Ppadog"),
    to = c("Pathway", "NGenes", "MeanAbsT0", "MeanWeightT0", "PValue", "FDR"),
    x = colnames(padog_result))



padog_result$items <- gs[match(padog_result$Name, names(gs))]
get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"])
}

se_comparison = se_comparison[igenes,]
padog_result$afc <- unlist(lapply(padog_result$items, function(x){get_AFC(se_comparison, x)}))

padog_result



plot(padog_result$afc, -log10(padog_result$PValue))
print(sum(padog_result$FDR<0.01, na.rm = T))

```


```{r With proteins and the md gmt}


upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Protein"
method = "padog"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)



#filter assay_data to remove not mapped proteins
#unmapped_proteins = read.csv("/home/joseph/experiments/ReactomeHERResults/not_found.csv", stringsAsFactors = F)
#unmapped_proteins$Not.found
#unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
#assay(se_comparison) = assay(se_comparison)[unmapped_filter,]

# perform normalization
normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
assay(se_comparison) = normalized_data

expression.data = assay(se_comparison)

# get gene sets
#gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP
padog_group <- rep(NA, ncol(expression.data))
padog_group[se_comparison$GROUP == 0] <- "c"
padog_group[se_comparison$GROUP == 1] <- "d"
padog_group


print("not using real padog group, using this instead")
padog_group = sample(padog_group)
print(padog_group)
# restrict gene sets and summarized experiment to interacting genes
# restrict se and gs to intersecting genes
igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                            "gene sets (gs) have no gene IDs in common")
expression.data <- expression.data[igenes,]
gs <- lapply(gs, function(s) s[s %in% igenes])

print("randomizing rows as well")
rownames(expression.data) <- sample(rownames(expression.data))

# run PADOG
padog_result <- padog(
    esetm = expression.data,
    group = padog_group,
    paired = FALSE,
    block = NULL,
    gslist = gs,
    NI = 1000, # number of iterations
    plots = FALSE,
    Nmin = 0) # minimum gene set size

# convert to the required result output
colnames(padog_result) <- plyr::mapvalues(
    from = c("ID", "Size", "meanAbsT0", "padog0", "PmeanAbsT", "Ppadog"),
    to = c("Pathway", "NGenes", "MeanAbsT0", "MeanWeightT0", "PValue", "FDR"),
    x = colnames(padog_result))



padog_result$items <- gs[match(padog_result$Name, names(gs))]
get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"])
}

se_comparison = se_comparison[igenes,]
padog_result$afc <- unlist(lapply(padog_result$items, function(x){get_AFC(se_comparison, x)}))

padog_result



plot(padog_result$afc, -log10(padog_result$PValue))
print(sum(padog_result$FDR<0.01, na.rm = T))

```











