---
title: "Benchmarking2"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(jsonlite)
library(EnrichmentAnalysisStepR)
```


# Check identifier mappings

Key learnings:
- isoform entries get duplicated so I need to account for this locally along with proteins that weren't mapped. 


```{r}
mapping = read.csv("/home/joseph/experiments/ReactomeGSAOutputs/HERResultsFinal/mapping.csv", stringsAsFactors = F)
mapping

unmapping = read.csv("/home/joseph/experiments/ReactomeGSAOutputs/HERResultsFinal/not_found.csv", stringsAsFactors = F)
unmapping

reactome.results = readxl::read_xlsx("/home/joseph/experiments/ReactomeGSAOutputs/HERResultsFinal/2c960850-c433-11eb-a876-fad8997f6386.xlsx")
reactome.results


submitted.data = read.csv("development/her_study_reactome.tsv", sep = "\t")
submitted.data
```

```{r}
length((submitted.data$ProteinId))
length(unique(submitted.data$ProteinId))
length((mapping$Submitted.identifier))
length(unique(mapping$Submitted.identifier))

# Reactome has removed rows
length(intersect(submitted.data$ProteinId, mapping$Submitted.identifier))
length(intersect(submitted.data$ProteinId, unmapping$Not.found))
sum(length(intersect(submitted.data$ProteinId, mapping$Submitted.identifier)) + 
    length(intersect(submitted.data$ProteinId, unmapping$Not.found)))
length(submitted.data$ProteinId)


# none of the rows are duplicated
sum(duplicated(submitted.data[,-1]))


# maybe reactome is removing proteins that aren't in the gs?
gmt.file = "/home/joseph/experiments/ReactomeGMT/ReactomeGMTTest.gmt"
gs <- getGenesets(gmt.file)

length(intersect(submitted.data$ProteinId, unlist(gs)))

# oh that's actually very close
length(intersect(intersect(submitted.data$ProteinId, 
                           mapping$Submitted.identifier), unlist(gs)))
# So they do remove proteins not in the gene set library before they list
# mapped and unmapped. 
# But we have 3 left? 

in_gs_and_mapped = (intersect(intersect(submitted.data$ProteinId, mapping$Submitted.identifier), unlist(gs)))


```

```{r Try to visualize overlaps}
library(ComplexHeatmap)
lt = list(
  submitted = submitted.data$ProteinId,
  mapped = mapping$Submitted.identifier,
  not_mapped = unmapping$Not.found
  #set_library = unlist(gs)
)

m = make_comb_mat(lt)
UpSet(m, comb_order = order(comb_size(m)))
```

#Compare Results: 
## PADOG

```{r Randomized Data}


upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Gene"
method = "padog"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)



#filter assay_data to remove not mapped proteins
unmapped_proteins = read.csv("/home/joseph/experiments/ReactomeHERResults/not_found.csv", stringsAsFactors = F)
unmapped_proteins$Not.found
unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
assay(se_comparison) = assay(se_comparison)[unmapped_filter,]

# perform normalization
normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
assay(se_comparison) = normalized_data

expression.data = assay(se_comparison)

# get gene sets
gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
#gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP
padog_group <- rep(NA, ncol(expression.data))
padog_group[se_comparison$GROUP == 0] <- "c"
padog_group[se_comparison$GROUP == 1] <- "d"
padog_group
print("not using real padog group, using this instead")
padog_group = sample(padog_group)
print(padog_group)
# restrict gene sets and summarized experiment to interacting genes
# restrict se and gs to intersecting genes
igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                            "gene sets (gs) have no gene IDs in common")
expression.data <- expression.data[igenes,]
gs <- lapply(gs, function(s) s[s %in% igenes])

print("randomizing rows as well")
rownames(expression.data) <- sample(rownames(expression.data))

# run PADOG
padog_result <- padog(
    esetm = expression.data,
    group = padog_group,
    paired = FALSE,
    block = NULL,
    gslist = gs,
    NI = 1000, # number of iterations
    plots = FALSE,
    Nmin = 0) # minimum gene set size

# convert to the required result output
colnames(padog_result) <- plyr::mapvalues(
    from = c("ID", "Size", "meanAbsT0", "padog0", "PmeanAbsT", "Ppadog"),
    to = c("Pathway", "NGenes", "MeanAbsT0", "MeanWeightT0", "PValue", "FDR"),
    x = colnames(padog_result))



padog_result$items <- gs[match(padog_result$Name, names(gs))]

get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"])
}

se_comparison = se_comparison[igenes,]
padog_result$afc <- unlist(lapply(padog_result$items, function(x){get_AFC(se_comparison, x)}))

plot(padog_result$afc, -log10(padog_result$PValue))
print(sum(padog_result$FDR<0.01, na.rm = T))
```

```{r Non-Randomized Data}

upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Protein"
method = "padog"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)



#filter assay_data to remove not mapped proteins
unmapped_proteins = read.csv("/home/joseph/experiments/ReactomeHERResults/not_found.csv", stringsAsFactors = F)
unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
se_comparison  = se_comparison[unmapped_filter,]

# perform normalization
#normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
#assay(se_comparison) = normalized_data

# get gene sets
gmt.file = "/home/joseph/experiments/ReactomeGMT/ReactomeGMTTest.gmt"
#gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
#gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP
padog_group <- rep(NA, ncol(expression.data))
padog_group[se_comparison$GROUP == 0] <- "c"
padog_group[se_comparison$GROUP == 1] <- "d"
padog_group

# restrict gene sets and summarized experiment to interacting genes
# restrict se and gs to intersecting genes
igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                            "gene sets (gs) have no gene IDs in common")
expression.data <- expression.data[igenes,]
gs <- lapply(gs, function(s) s[s %in% igenes])


# run PADOG
padog_result <- padog(
    esetm = expression.data,
    group = padog_group,
    paired = FALSE,
    block = NULL,
    gslist = gs,
    NI = 1000, # number of iterations
    plots = FALSE,
    Nmin = 0) # minimum gene set size

# convert to the required result output
colnames(padog_result) <- plyr::mapvalues(
    from = c("ID", "Size", "meanAbsT0", "padog0", "PmeanAbsT", "Ppadog"),
    to = c("Pathway", "NGenes", "MeanAbsT0", "MeanWeightT0", "PValue", "FDR"),
    x = colnames(padog_result))



padog_result$items <- gs[match(padog_result$Name, names(gs))]

get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"])
}

se_comparison = se_comparison[igenes,]
padog_result$afc <- unlist(lapply(padog_result$items, function(x){get_AFC(se_comparison, x)}))

plot(padog_result$afc, -log10(padog_result$PValue))
print(sum(padog_result$FDR<0.01, na.rm = T))
```

```{r add name of pathways }
# use annotation file to get names
annotations = read.csv("/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome_set_info.csv", stringsAsFactors = F)
colnames(annotations) = c("Pathway","Name", "description","items","count","link")
padog_result <- merge(padog_result, annotations, by = "Pathway")
```

```{r Benchmark Against Server Results when using padog directly/min size set to 0}
md_results = padog_result

reactome_path = "/home/joseph/experiments/ReactomeGSAOutputs/ReactomeHERResults/padog_results.xlsx"
reactome_results = readxl::read_excel(reactome_path)

plot(reactome_results$`av_foldchange.HER2 Resistance Study`, -log10(reactome_results$`PValue.HER2 Resistance Study`))
plot(-md_results$afc, -log10(md_results$PValue))


both_results = merge(reactome_results, md_results, by.x = "Name", by.y = "Name.y")

plot(both_results$`av_foldchange.HER2 Resistance Study`, -both_results$afc)
plot(-log10(both_results$`PValue.HER2 Resistance Study`), 
     -log10(both_results$PValue))
plot(both_results$`NGenes.HER2 Resistance Study`, both_results$NGenes)
abline(0,1)

cor(both_results$`av_foldchange.HER2 Resistance Study`, -both_results$afc)
cor(-log10(both_results$`PValue.HER2 Resistance Study`), -log10(both_results$PValue))

tmp = both_results[both_results$NGenes>0,]
sum((tmp$`PValue.HER2 Resistance Study`<0.05) & (tmp$PValue <0.05))
sum((tmp$`PValue.HER2 Resistance Study`<0.05) & (tmp$PValue >0.05))
sum((tmp$`PValue.HER2 Resistance Study`>0.05) & (tmp$PValue <0.05))
sum((tmp$`PValue.HER2 Resistance Study`>0.05) & (tmp$PValue >0.05))


```

```{r}
both_results[both_results$NGenes==0,]
```

```{r}
missing_proteins = c('Q8WU20', 'O43559', 'P34130', 'Q16620', 'P01112', 'P01111', 'Q06124', 'P23560', 'Q07889')
intersect(submitted.data$ProteinId, missing_proteins)

```

```{r}
"P62993" %in% submitted.data$ProteinId
```

```{r}
unmapped_proteins = read.csv("/home/joseph/experiments/ReactomeHERResults/not_found.csv", stringsAsFactors = F)
mapped_proteins = read.csv("/home/joseph/experiments/ReactomeHERResults/mapping.csv", stringsAsFactors = F)

dim(unmapped_proteins)[1] + dim(mapped_proteins)[1]

dim(assay(se_comparison))


rownames(se_comparison)
```

```{r}
assay(se_comparison)[c("Q14103","Q14103"),]
```

```{r}
"Q14103" %in% unlist(gs)
"Q14103-3" %in% unlist(gs)
"Q15717" %in% unlist(gs)

mapped = []
se_comparison[mapped_proteins$Found.identifier,]
```

```{r}
length(intersect(rownames(se_comparison), mapped_proteins$Submitted.identifier))
length(rownames(se_comparison))


tmp = mapped_proteins[mapped_proteins$Submitted.identifier %in% rownames(se),]

se_comparison[tmp$Submitted.identifier]

```


## Camera

```{r Randomized Data}
upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Gene"
method = "Camera"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)



#filter assay_data to remove not mapped proteins
#unmapped_proteins = read.csv("/home/joseph/experiments/ReactomeHERResults/not_found.csv", stringsAsFactors = F)
#unmapped_proteins$Not.found
#unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
#assay(se_comparison) = assay(se_comparison)[unmapped_filter,]

# perform normalization
normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
assay(se_comparison) = normalized_data

expression.data = assay(se_comparison)

# get gene sets
gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
#gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP

### Column randomization
print("not using real padog group, using this instead")
se_comparison$GROUP <- sample(se_comparison$GROUP)
se_comparison$GROUP

# restrict gene sets and summarized experiment to interacting genes
# restrict se and gs to intersecting genes
igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                            "gene sets (gs) have no gene IDs in common")
expression.data <- expression.data[igenes,]
gs <- lapply(gs, function(s) s[s %in% igenes])



#### Row randomization
print("randomizing rows as well")
rownames(se_comparison) <- sample(rownames(se_comparison))


### USE CAMERA HERE

## verbatim code from GSA/sbea (at first)
se = se_comparison

# design matrix
grp <- colData(se)[, configEBrowser("GRP.COL")]
blk <- NULL
BLK.COL <- configEBrowser("BLK.COL")
if(BLK.COL %in% colnames(colData(se))) blk <- colData(se)[,BLK.COL]

group <- factor(grp)
paired <- !is.null(blk)
f <- "~" 
if(paired) 
{   
  block <- factor(blk)
  f <- paste0(f, "block + ") 
}   
f <- formula(paste0(f, "group"))
design <- model.matrix(f)

y <- assay(se)

# set gene sets
gs.index <- limma::ids2indices(gs, rownames(se))

# run roast / camera
camera.result <- limma::camera(y, gs.index, design, sort=FALSE)

# formal results
gsa.result <- camera.result[, c("Direction", "FDR", "PValue", "NGenes")]
gsa.result$Pathway <- rownames(gsa.result)


gsa.result$items <- gs[match(names(gsa.result), names(gs))]


get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"])
}

se_comparison = se_comparison[igenes,]
#gsa.result$afc <- unlist(lapply(gsa.result$items, function(x){get_AFC(se_comparison, x)}))


sum(gsa.result$PValue<0.05)
sum(gsa.result$FDR<0.05)
hist(gsa.result$PValue)


gsa.result[gsa.result$PValue<0.05,]$FDR
```

```{r Real Experiment Gene}
upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Gene"
method = "camera"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)



#filter assay_data to remove not mapped proteins
#unmapped_proteins = read.csv("/home/joseph/experiments/ReactomeHERResults/not_found.csv", stringsAsFactors = F)
#unmapped_proteins$Not.found
#unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
#assay(se_comparison) = assay(se_comparison)[unmapped_filter,]

# perform normalization
normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
assay(se_comparison) = normalized_data

expression.data = assay(se_comparison)

# get gene sets
gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
#gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP

### Column randomization
print("No column randomization")
#se_comparison$GROUP <- sample(se_comparison$GROUP)
#se_comparison$GROUP

# restrict gene sets and summarized experiment to interacting genes
# restrict se and gs to intersecting genes
igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                            "gene sets (gs) have no gene IDs in common")
expression.data <- expression.data[igenes,]
gs <- lapply(gs, function(s) s[s %in% igenes])



#### Row randomization
print("Not randomizing rows")
#rownames(se_comparison) <- sample(rownames(se_comparison))


### USE CAMERA HERE

## verbatim code from GSA/sbea (at first)
se = se_comparison

# design matrix
grp <- colData(se)[, configEBrowser("GRP.COL")]
blk <- NULL
BLK.COL <- configEBrowser("BLK.COL")
if(BLK.COL %in% colnames(colData(se))) blk <- colData(se)[,BLK.COL]

group <- factor(grp)
paired <- !is.null(blk)
f <- "~" 
if(paired) 
{   
  block <- factor(blk)
  f <- paste0(f, "block + ") 
}   
f <- formula(paste0(f, "group"))
design <- model.matrix(f)

y <- assay(se)

# set gene sets
gs.index <- limma::ids2indices(gs, rownames(se))

# run roast / camera
camera.result <- limma::camera(y, gs.index, design, sort=FALSE)
colnames(camera.result)
# formal results
gsa.result <- camera.result[, c("Direction", "FDR", "PValue", "NGenes")]
gsa.result$Pathway <- rownames(gsa.result)


gsa.result$items <- gs[match(rownames(gsa.result), names(gs))]


get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"])
}

se_comparison = se_comparison[igenes,]
gsa.result$afc <- unlist(lapply(gsa.result$items, function(x){get_AFC(se_comparison, x)}))


sum(gsa.result$PValue<0.05)
sum(gsa.result$FDR<0.05)
hist(gsa.result$PValue)

```

```{r Real Experiment Protein}
upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Protein"
method = "camera"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)



#filter assay_data to remove not mapped proteins
#unmapped_proteins = read.csv("/home/joseph/experiments/ReactomeHERResults/not_found.csv", stringsAsFactors = F)
#unmapped_proteins$Not.found
#unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
#assay(se_comparison) = assay(se_comparison)[unmapped_filter,]

# perform normalization
normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
#assay(se_comparison) = normalized_data

expression.data = assay(se_comparison)

# get gene sets
#gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP

### Column randomization
print("No column randomization")
#se_comparison$GROUP <- sample(se_comparison$GROUP)
#se_comparison$GROUP


if (F){
  # restrict gene sets and summarized experiment to interacting genes
  # restrict se and gs to intersecting genes
  igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
  if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                           "gene sets (gs) have no gene IDs in common")
  expression.data <- expression.data[igenes,]
  gs <- lapply(gs, function(s) s[s %in% igenes])
  
  
  
}


#### Row randomization
print("Not randomizing rows")
#rownames(se_comparison) <- sample(rownames(se_comparison))


### USE CAMERA HERE

## verbatim code from GSA/sbea (at first)
se = se_comparison

# design matrix
grp <- colData(se)[, configEBrowser("GRP.COL")]
blk <- NULL
BLK.COL <- configEBrowser("BLK.COL")
if(BLK.COL %in% colnames(colData(se))) blk <- colData(se)[,BLK.COL]

group <- factor(grp)
paired <- !is.null(blk)
f <- "~" 
if(paired) 
{   
  block <- factor(blk)
  f <- paste0(f, "block + ") 
}   
f <- formula(paste0(f, "group"))
design <- model.matrix(f)

y <- expression.data

# set gene sets
gs.index <- limma::ids2indices(gs, rownames(y))

# run roast / camera
camera.result <- limma::camera(y, gs.index, design, sort=FALSE)
colnames(camera.result)
# formal results
gsa.result <- camera.result[, c("Direction", "FDR", "PValue", "NGenes")]
gsa.result$Pathway <- rownames(gsa.result)


gsa.result <- merge(gsa.result, gs2df(gs), by.x = "Pathway", by.y = "gene.set")


get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"])
}


gs2df <- function(gs){
  df <- as.data.frame(cbind(names(gs),gs))
  colnames(df) <- c("gene.set","items")
  df$gene.set <- sapply(df$gene.set, toString)
  df
}

#se_comparison = se_comparison[igenes,]
gsa.result$afc <- unlist(lapply(gsa.result$items, function(x){get_AFC(se_comparison, x)}))

sum(gsa.result$PValue<0.05)
sum(gsa.result$FDR<0.05)
hist(gsa.result$PValue)

```

```{r}
# use annotation file to get names
annotations = read.csv("/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome_set_info.csv", stringsAsFactors = F)
colnames(annotations) = c("Pathway","Name", "description","items","count","link")
gsa.result <- merge(gsa.result, annotations, by = "Pathway")

```

```{r Import Reactome Camera Results and Compare}
  camera_results_her = readxl::read_xlsx("/home/joseph/experiments/ReactomeGSAOutputs/ReactomeHERResultsCamera/cb807b4c-c34f-11eb-a235-3e4fca4d787e.xlsx")

colnames(camera_results_her)
colnames(gsa.result)
head(gsa.result)
head(camera_results_her)

# proteins
both_results = merge(gsa.result, camera_results_her, by = "Name")
# genes 
# both_results = merge(gsa.result, camera_results_her, by.x = "Pathway", by.y = "Name")


plot(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
plot(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$PValue))

cor(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
cor(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
cor(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$PValue))

```

## Reactome GSA comparison

```{r}
library(ReactomeGSA.data)

data(griss_melanoma_proteomics)

griss_melanoma_proteomics$E

```

# Create Isoform Mapped version of the local data


```{r}
upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Protein"
method = "padog"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)
```

```{r}
se_comparison  = se_comparison[match(mapping$Submitted.identifier, rownames(se_comparison)),]
sum(duplicated(rownames(se_comparison)))
rownames(se_comparison) <- mapping$Found.identifier
sum(duplicated(rownames(to_submit)))
```

```{r Try to run enrichment using mapping dataframe}
upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Protein"

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)


# Get Mappings
mapping = read.csv("/home/joseph/experiments/ReactomeGSAOutputs/HERResultsFinal/mapping.csv", stringsAsFactors = F)
mapping

unmapping = read.csv("/home/joseph/experiments/ReactomeGSAOutputs/HERResultsFinal/not_found.csv", stringsAsFactors = F)
unmapping

reactome.results = readxl::read_xlsx("/home/joseph/experiments/ReactomeGSAOutputs/HERResultsFinal/2c960850-c433-11eb-a876-fad8997f6386.xlsx")
reactome.results


submitted.data = read.csv("development/her_study_reactome.tsv", sep = "\t")
submitted.data

# Filter for data that was mapped
se_comparison  = se_comparison[match(mapping$Submitted.identifier,rownames(se_comparison)),]
sum(duplicated(rownames(se_comparison))) # added rows
rownames(se_comparison) <- mapping$Found.identifier
sum(duplicated(rownames(se_comparison))) # fixed identifiers

# get gene sets
gmt.file = "/home/joseph/experiments/ReactomeGMT/ReactomeGMTTest.gmt"
#gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
#gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP
padog_group <- rep(NA, ncol(se_comparison))
padog_group[se_comparison$GROUP == 0] <- "c"
padog_group[se_comparison$GROUP == 1] <- "d"
padog_group

# restrict gene sets and summarized experiment to interacting genes
# restrict se and gs to intersecting genes
igenes <- intersect(rownames(se_comparison), unique(unlist(gs)))
if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                            "gene sets (gs) have no gene IDs in common")
se_comparison <- se_comparison[igenes,]
gs <- lapply(gs, function(s) s[s %in% igenes])


# run PADOG
padog_result <- padog(
    esetm = assay(se_comparison),
    group = padog_group,
    paired = FALSE,
    block = NULL,
    gslist = gs,
    NI = 1000, # number of iterations
    plots = FALSE,
    Nmin = 0) # minimum gene set size

# convert to the required result output
colnames(padog_result) <- plyr::mapvalues(
    from = c("ID", "Size", "meanAbsT0", "padog0", "PmeanAbsT", "Ppadog"),
    to = c("Pathway", "NGenes", "MeanAbsT0", "MeanWeightT0", "PValue", "FDR"),
    x = colnames(padog_result))


padog_result$items <- gs[match(padog_result$Name, names(gs))]

get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"])
}

se_comparison = se_comparison[igenes,]
padog_result$afc <- unlist(lapply(padog_result$items, function(x){get_AFC(se_comparison, x)}))

plot(padog_result$afc, -log10(padog_result$PValue))
print(sum(padog_result$FDR<0.01, na.rm = T))
```


```{r add name of pathways }
# use annotation file to get names
annotations = read.csv("/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome_set_info.csv", stringsAsFactors = F)
colnames(annotations) = c("Pathway","Name", "description","items","count","link")
padog_result <- merge(padog_result, annotations, by = "Pathway")
```

```{r Benchmark Against Server Results when using padog directly/min size set to 0}
md_results = padog_result

reactome_path = "/home/joseph/experiments/ReactomeGSAOutputs/ReactomeHERResults/padog_results.xlsx"
reactome_results = readxl::read_excel(reactome_path)

plot(reactome_results$`av_foldchange.HER2 Resistance Study`, -log10(reactome_results$`PValue.HER2 Resistance Study`))
plot(-md_results$afc, -log10(md_results$PValue))


both_results = merge(reactome_results, md_results, by.x = "Name", by.y = "Name.y")

plot(both_results$`av_foldchange.HER2 Resistance Study`, -both_results$afc)
plot(both_results$`MeanAbsT0.HER2 Resistance Study`, both_results$MeanAbsT0)
plot(both_results$`MeanWeightT0.HER2 Resistance Study`, both_results$MeanWeightT0)


plot(-log10(both_results$`FDR.HER2 Resistance Study`), 
     -log10(both_results$FDR))
plot(both_results$`NGenes.HER2 Resistance Study`, both_results$NGenes)
abline(0,1)

cor(both_results$`av_foldchange.HER2 Resistance Study`, -both_results$afc)
cor(-log10(both_results$`PValue.HER2 Resistance Study`), -log10(both_results$PValue))

tmp = both_results[both_results$NGenes>0,]
sum((tmp$`PValue.HER2 Resistance Study`<0.05) & (tmp$PValue <0.05))
sum((tmp$`PValue.HER2 Resistance Study`<0.05) & (tmp$PValue >0.05))
sum((tmp$`PValue.HER2 Resistance Study`>0.05) & (tmp$PValue <0.05))
sum((tmp$`PValue.HER2 Resistance Study`>0.05) & (tmp$PValue >0.05))


```

```{r}
both_results[both_results$NGenes != both_results$`NGenes.HER2 Resistance Study`,
             c("Name", "NGenes", "NGenes.HER2 Resistance Study")]
```



# Reactome Benchmark Attempt 3:

```{r Real Experiment Protein}
upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Protein"
method = "camera"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)

#filter assay_data to remove not mapped proteins
unmapped_proteins = read.csv("/home/joseph/experiments/ReactomeGSAOutputs/ReactomeHERResultsCamera2/not_found.csv", stringsAsFactors = F)
unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
se_comparison = se_comparison[unmapped_filter, ]

# perform normalization
#normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
#assay(se_comparison) = normalized_data

# get gene sets
# protein centric copy of reactome protein gmt
gmt.file = "/home/joseph/experiments/ReactomeGMT/ReactomeGMTTest.gmt"
#gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
#gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP


if (F){
  # restrict gene sets and summarized experiment to interacting genes
  # restrict se and gs to intersecting genes
  igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
  if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                           "gene sets (gs) have no gene IDs in common")
  expression.data <- expression.data[igenes,]
  gs <- lapply(gs, function(s) s[s %in% igenes])
}

### USE CAMERA HERE

## verbatim code from GSA/sbea (at first)
se = se_comparison

# design matrix
grp <- colData(se)[, configEBrowser("GRP.COL")]
blk <- NULL
BLK.COL <- configEBrowser("BLK.COL")
if(BLK.COL %in% colnames(colData(se))) blk <- colData(se)[,BLK.COL]

group <- factor(grp)
paired <- !is.null(blk)
f <- "~" 
if(paired) 
{   
  block <- factor(blk)
  f <- paste0(f, "block + ") 
}   
f <- formula(paste0(f, "group"))
design <- model.matrix(f)

y <- assay(se_comparison)

# set gene sets

# traditional way
#gs.index <- limma::ids2indices(gs, rownames(y))

# reactome gsa method
mapping = read.csv("/home/joseph/experiments/ReactomeGSAOutputs/ReactomeHERResultsCamera2/mapping.csv", stringsAsFactors = F)
identifier_mapping <- mapping$Found.identifier
names(identifier_mapping) <- mapping$Submitted.identifier

gs.index <- ids2indices_multiple(identifier_mapping, gs, y)

# run roast / camera
camera.result <- limma::camera(y, gs.index, design, sort=FALSE)
colnames(camera.result)
# formal results
gsa.result <- camera.result[, c("Direction", "FDR", "PValue", "NGenes")]
gsa.result$Pathway <- rownames(gsa.result)


gsa.result <- merge(gsa.result, gs2df(gs), by.x = "Pathway", by.y = "gene.set")

get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"], na.rm = T)
}

gs2df <- function(gs){
  df <- as.data.frame(cbind(names(gs),gs))
  colnames(df) <- c("gene.set","items")
  df$gene.set <- sapply(df$gene.set, toString)
  df
}

#se_comparison = se_comparison[igenes,]
gsa.result$afc <- unlist(lapply(gsa.result$items, function(x){get_AFC(se_comparison, x)}))
gsa.result <- add_pathway_foldchanges(gsa.result, fold_changes, gs.index, assay(se_comparison))

sum(gsa.result$PValue<0.05)
sum(gsa.result$FDR<0.05)
hist(gsa.result$PValue)

```

```{r Assign Pathway Names in GSA object}
# use annotation file to get names
annotations = read.csv("/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome_set_info.csv", stringsAsFactors = F)
colnames(annotations) = c("Pathway","Name", "description","items","count","link")
gsa.result <- merge(gsa.result, annotations, by = "Pathway")

```

```{r Import Reactome Camera Results and Compare}
camera_results_her = readxl::read_xlsx("/home/joseph/experiments/ReactomeGSAOutputs/ReactomeHERResultsCamera2/5b56bc62-c9ae-11eb-bc4d-cae9afd38b60.xlsx")

colnames(camera_results_her)
colnames(gsa.result)
head(gsa.result)
head(camera_results_her)

# proteins
both_results = merge(gsa.result, camera_results_her, by = "Name")
# genes 
# both_results = merge(gsa.result, camera_results_her, by.x = "Pathway", by.y = "Name")

plot(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$av_foldchange)
plot(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$PValue))

cor(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
cor(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
cor(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$PValue))

```

```{r}
tmp = both_results[both_results$NGenes != both_results$NGenes.PROTEOMICS_INT_1,]
tmp[,c(2,6,16)]

tmp$NGenes[[1]]
tmp$NGenes.PROTEOMICS_INT_1[[1]]
tmp$items.x[[1]]

gs[[tmp$Pathway[[1]]]]

intersect(rownames(se_comparison), gs[[tmp$Pathway[[1]]]])

length(tmp$NGenes - tmp$NGenes.PROTEOMICS_INT_1)
tmp$NGenes - tmp$NGenes.PROTEOMICS_INT_1
mean(tmp$NGenes - tmp$NGenes.PROTEOMICS_INT_1)
```

```{r Double Check AFC}
colnames(tmp)
tmp[2, c("Pathway", "afc", "av_foldchange.PROTEOMICS_INT_1", "items.x", "NGenes", "NGenes.PROTEOMICS_INT_1")]
proteins <- intersect(unlist(tmp[2,"items.x"]), rownames(assay(se_comparison)))
mean(rowData(se_comparison)[proteins,"FC"])
```

```{r Does anything correlate specifically with the different p-values}
library(plotly)
library(ggplot2)

both_results$Gene_Same = both_results$NGenes == both_results$NGenes.PROTEOMICS_INT_1
ggplot(both_results, 
       aes(PValue, PValue.PROTEOMICS_INT_1, color = NGenes)) + 
  geom_point() +
  scale_x_continuous(trans='log10') +
  scale_y_continuous(trans='log10') 


```

```{r Can we identify the different proteins?}
tmp = both_results[both_results$NGenes != both_results$NGenes.PROTEOMICS_INT_1,]
reverse_tmp = both_results[both_results$NGenes == both_results$NGenes.PROTEOMICS_INT_1,]
proteins_in_sets_with_dif_ngenes = intersect(unlist(tmp$items.x), 
                                             rownames(assay(se_comparison)))
proteins_in_sets_with_same_ngenes = intersect(unlist(reverse_tmp$items.x), 
                                             rownames(assay(se_comparison)))

proteins_in_both = intersect(proteins_in_sets_with_dif_ngenes, 
                             proteins_in_sets_with_same_ngenes)

proteins_just_in_dif_ngene_sets <- setdiff(proteins_in_sets_with_dif_ngenes, proteins_in_both)
length(proteins_just_in_dif_ngene_sets)


filtered_assay <- assay(se_comparison)[!(rownames(assay(se_comparison)) %in% proteins_just_in_dif_ngene_sets),]
dim(filtered_assay)

filter
write.table(filtered_assay, "filtered_her_assay.tsv",  sep ="\t", quote = F, row.names = T)

```
## utilities

```{r}

#' Replicates \code{\link[limma]{ids2indices}} while supporting multiple mappings
#'
#' When mapping identifiers, some identifiers often map to multiple entries in the target namespace.
#' This function replicates \code{\link[limma]{ids2indices}} but adds support for these multiple mappings.
#'
#' If a protein / gene maps to multiple entries in the target namespace, the gene / protein will be referenced
#' once in every pathway where any of these mapped entries occurs.
#'
#' @param identifier_mapping A named list with the original gene/protein identifier as key and all mappings as
#'        a vector
#' @param gene_set A named list with pathway ids as names and their member genes / proteins as vactors
#' @param expression.data \code{data.frame} containing the expression data for genes / proteins (rows) per sample
#'        (column)
ids2indices_multiple <- function(identifier_mapping, gene_set, expression.data) {
    # make sure the identifier_mapping has the same sort order as the rows in the expression.data
    identifier_mapping <- identifier_mapping[rownames(expression.data)]

    # map every identifier to a position
    identifier_position = c()
    for (i in 1:length(identifier_mapping)) {
        identifier_position[identifier_mapping[[i]]] = i
    }

    # get the gene position per gene set
    gene.indices <- lapply(gene_set, function(x) {
        existing.genes <- x[x %in% names(identifier_position)]
        positions <- as.numeric(identifier_position[existing.genes])
        positions <- unique(positions)
        return(positions)
    })

    # only keep gene sets with mappings
    gene.indices <- gene.indices[lengths(gene.indices) != 0L]

    return(gene.indices)
}

  
#' Removes unmapped genes
#'
#' Removes all genes/proteins from the expression matrix that do not appear in the mapping result ie. that
#' do not correspond to names in the \code{identifier_mapping} list.
#'
#' @param expression.data data.frame containing genes in rows and samples in columns
#' @param identifier_mapping names list with the genes as names and their mappings as values
remove_unmapped <- function(expression.data, identifier_mapping) {
    expression.data <- expression.data[rownames(expression.data) %in% names(identifier_mapping), ]

    return(expression.data)
}

#' Add average fold changes for pathway result
#'
#' @param result Result of the GSA algorithm as a data.frame
#' @param fold_changes Molecule level fold_changes
#' @param gene_index Index of genes mapping to a pathway
#' @param expression_data A data.frame containing the expression matrix
#' @return An updated version of result with an additional column 'av_foldchange'
add_pathway_foldchanges <- function(result, fold_changes, gene_index, expression_data) {
    # make sure the fold_changes contain the required columns
    for (required_col in c("Identifier", "logFC")) {
        if (!required_col %in% colnames(fold_changes)) {
            stop("Error: Missing ", required_col, " in fold_changes: ", paste(colnames(fold_changes)))
        }
    }

    # change the fold_changes to use the identifier as rownames
    rownames(fold_changes) <- fold_changes$Identifier

    # calculate the average fold-change for every pathway
    av_fc <- sapply(result$Pathway, function(pathway_id) {
        # get all genes for that pathway
        if (!pathway_id %in% names(gene_index)) {
            stop("Error: Missing index for ", pathway_id)
        }

        gene_ids <- rownames(expression_data)[gene_index[[pathway_id]]]

        this_av_fc <- mean(fold_changes[gene_ids, "logFC"], na.rm = T)

        # return the average logFC
        return(this_av_fc)
    })

    # add the average fold-changes as a new column
    result$av_foldchange <- as.numeric(av_fc)

    # add the direction if it isn't present
    if (!"Direction" %in% colnames(result)) {
        org_colnames <- colnames(result)
        result$Direction <- ifelse(result$av_foldchange > 0, "Up", "Down")

        result <- result[, c(org_colnames[1:2], "Direction", org_colnames[3:length(org_colnames)])]
    }

    return(result)
}
```

```{r creating mapping_input}
mapping = read.csv("/home/joseph/experiments/ReactomeGSAOutputs/HERResultsFinal/mapping.csv", stringsAsFactors = F)
identifier_mapping <- mapping$Found.identifier
names(identifier_mapping) <- mapping$Submitted.identifier
gene.indices <- ids2indices_multiple(identifier_mapping, gs, assay_data)
```


```{r}
gsa.result
```

```{r test fold changes code}
fold_changes <- rowData(se_comparison)
colnames(fold_changes)
colnames(fold_changes) <- c("Identifier", "GeneName","PVAL","ADJ.PVAL", "logFC")

result <- add_pathway_foldchanges(gsa.result, fold_changes, gs.index, assay(se_comparison))

colnames(result)
```


```{r test fold changes code}
gene.indices

```



# Reactome Benchmark Attempt 4:

This iterations:
- GMT from iteration 3
- Index mapping and afc from reactome GSA
- Filtered Data to attempt to remove discrepancies in GMT
(issue sidestep)

- CAMERA
- no normalization


```{r Real Experiment Protein}
upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Protein"
method = "camera"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)


protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)

#filter assay_data to remove not mapped proteins
unmapped_proteins = read.csv("/home/joseph/experiments/ReactomeGSAOutputs/ReactomeHERResultsFilteredCamera/not_found.csv", stringsAsFactors = F)
unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
se_comparison = se_comparison[unmapped_filter, ]

# perform normalization
#normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
#assay(se_comparison) = normalized_data

# get gene sets
# protein centric copy of reactome protein gmt
gmt.file = "/home/joseph/experiments/ReactomeGMT/ReactomeGMTTest.gmt"
#gmt.file = "/home/joseph/experiments/reactome_pathways_gene_sets/ReactomePathways.gmt/ReactomePathways.gmt"
#gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP


if (F){
  # restrict gene sets and summarized experiment to interacting genes
  # restrict se and gs to intersecting genes
  igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
  if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                           "gene sets (gs) have no gene IDs in common")
  expression.data <- expression.data[igenes,]
  gs <- lapply(gs, function(s) s[s %in% igenes])
}

### USE CAMERA HERE

## verbatim code from GSA/sbea (at first)
se = se_comparison

# design matrix
grp <- colData(se)[, configEBrowser("GRP.COL")]
blk <- NULL
BLK.COL <- configEBrowser("BLK.COL")
if(BLK.COL %in% colnames(colData(se))) blk <- colData(se)[,BLK.COL]

group <- factor(grp)
paired <- !is.null(blk)
f <- "~" 
if(paired) 
{   
  block <- factor(blk)
  f <- paste0(f, "block + ") 
}   
f <- formula(paste0(f, "group"))
design <- model.matrix(f)

y <- assay(se_comparison)

# set gene sets

# traditional way
#gs.index <- limma::ids2indices(gs, rownames(y))

# reactome gsa method
mapping = read.csv("/home/joseph/experiments/ReactomeGSAOutputs/ReactomeHERResultsFilteredCamera/mapping.csv", stringsAsFactors = F)
identifier_mapping <- mapping$Found.identifier
names(identifier_mapping) <- mapping$Submitted.identifier

gs.index <- ids2indices_multiple(identifier_mapping, gs, y)

# run roast / camera
camera.result <- limma::camera(y, gs.index, design, sort=FALSE)
colnames(camera.result)
# formal results
gsa.result <- camera.result[, c("Direction", "FDR", "PValue", "NGenes")]
gsa.result$Pathway <- rownames(gsa.result)


gsa.result <- merge(gsa.result, gs2df(gs), by.x = "Pathway", by.y = "gene.set")

get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"], na.rm = T)
}

gs2df <- function(gs){
  df <- as.data.frame(cbind(names(gs),gs))
  colnames(df) <- c("gene.set","items")
  df$gene.set <- sapply(df$gene.set, toString)
  df
}

#se_comparison = se_comparison[igenes,]
gsa.result$afc <- unlist(lapply(gsa.result$items, function(x){get_AFC(se_comparison, x)}))
gsa.result <- add_pathway_foldchanges(gsa.result, fold_changes, gs.index, assay(se_comparison))

sum(gsa.result$PValue<0.05)
sum(gsa.result$FDR<0.05)
hist(gsa.result$PValue)

```

```{r Assign Pathway Names in GSA object}
# use annotation file to get names
annotations = read.csv("/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome_set_info.csv", stringsAsFactors = F)
colnames(annotations) = c("Pathway","Name", "description","items","count","link")
gsa.result <- merge(gsa.result, annotations, by = "Pathway")

```

```{r Import Reactome Camera Results and Compare}
camera_results_her = readxl::read_xlsx("/home/joseph/experiments/ReactomeGSAOutputs/ReactomeHERResultsFilteredCamera/242f77da-cb71-11eb-a823-fad8997f6386.xlsx")

colnames(camera_results_her)
colnames(gsa.result)
head(gsa.result)
head(camera_results_her)

# proteins
both_results = merge(gsa.result, camera_results_her, by = "Name")
# genes 
# both_results = merge(gsa.result, camera_results_her, by.x = "Pathway", by.y = "Name")

plot(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$av_foldchange)
plot(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$PValue))

cor(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
cor(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
cor(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$PValue))

```

```{r}
tmp = both_results[both_results$NGenes != both_results$NGenes.PROTEOMICS_INT_1,]
tmp[,c(2,6,16)]

tmp$NGenes[[1]]
tmp$NGenes.PROTEOMICS_INT_1[[1]]
tmp$items.x[[1]]

gs[[tmp$Pathway[[1]]]]

intersect(rownames(se_comparison), gs[[tmp$Pathway[[1]]]])

length(tmp$NGenes - tmp$NGenes.PROTEOMICS_INT_1)
tmp$NGenes - tmp$NGenes.PROTEOMICS_INT_1
mean(tmp$NGenes - tmp$NGenes.PROTEOMICS_INT_1)
```

```{r Double Check AFC}
colnames(tmp)
tmp[2, c("Pathway", "afc", "av_foldchange.PROTEOMICS_INT_1", "items.x", "NGenes", "NGenes.PROTEOMICS_INT_1")]
proteins <- intersect(unlist(tmp[2,"items.x"]), rownames(assay(se_comparison)))
mean(rowData(se_comparison)[proteins,"FC"])
```

```{r Does anything correlate specifically with the different p-values}
library(plotly)
library(ggplot2)

both_results$Gene_Same = both_results$NGenes == both_results$NGenes.PROTEOMICS_INT_1
ggplot(both_results, 
       aes(PValue, PValue.PROTEOMICS_INT_1, color = NGenes)) + 
  geom_point() +
  scale_x_continuous(trans='log10') +
  scale_y_continuous(trans='log10') 


```

```{r}
plot(both_results$afc, both_results$av_foldchange)
plot(both_results$afc, -log10(both_results$FDR))
plot(both_results$av_foldchange, -log10(both_results$FDR))


p <- ggplot(both_results, 
            aes(av_foldchange, -log10(FDR), color = NGenes, gene.set = Pathway, name = Name)) + 
  geom_point() +
  geom_hline(yintercept=-log10(0.05)) 
ggplotly(p)

```



```{r Can we identify the different proteins?}
tmp = both_results[both_results$NGenes != both_results$NGenes.PROTEOMICS_INT_1,]
reverse_tmp = both_results[both_results$NGenes == both_results$NGenes.PROTEOMICS_INT_1,]
proteins_in_sets_with_dif_ngenes = intersect(unlist(tmp$items.x), 
                                             rownames(assay(se_comparison)))
proteins_in_sets_with_same_ngenes = intersect(unlist(reverse_tmp$items.x), 
                                             rownames(assay(se_comparison)))

proteins_in_both = intersect(proteins_in_sets_with_dif_ngenes, 
                             proteins_in_sets_with_same_ngenes)

proteins_just_in_dif_ngene_sets <- setdiff(proteins_in_sets_with_dif_ngenes, proteins_in_both)
length(proteins_just_in_dif_ngene_sets)


filtered_assay <- assay(se_comparison)[!(rownames(assay(se_comparison)) %in% proteins_just_in_dif_ngene_sets),]
dim(filtered_assay)

filter
write.table(filtered_assay, "filtered_her_assay.tsv",  sep ="\t", quote = F, row.names = T)

```