---
title: "ReactomeComparison"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("ReactomeGSAFunctions.R")
library(jsonlite)
library(EnrichmentAnalysisStepR)
```

# Part 1 Achieving Equivalency with Reactome

In order to achieve parallel results with Reactome when using the Mass Dynamics
Enrichment Feature, it was first necessary to align our methods with theirs in a
number of ways. 

The critical components in an any enrichment workflow are as follows:

Pre-processing of the experimental data:
- Removing missing values (dealing with missing values)
- Normalization (if required)
- specifying the design

Sourcing and Mapping to a Gene Set Library
- Selecting the origin of the annotations
- Generating a GMT file from an annotation file
- Mapping any identifiers (dealing with isoforms)

Processing
- Choosing an enrichment algorithm
- Specifying the enrichment parameters

Showing the Results:
- joining pathway name/size information with gsa results
- Deciding how to plot/visualize/summarize

In the process of achieving this comparison to Reactome, we had to solve many 
discrepancies (or at least eliminate them as possibilities) before we eventually
achieved a degree of parity (which was imperfect). 

# ReactomeGSA vs MD "ReactomeGSA Hybrid"


```{r}
upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Protein"
method = "camera"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)

protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)
```

```{r}
write.table(assay(se_comparison), "her_data_14_6_21.tsv",  sep ="\t", quote = F, row.names = T)
# I then manually add the column name for the index
```

![Parameters Used](/home/joseph/github_repos/EnrichmentAnalysisStepR/benchmarking/CallingReactomeGSA.PNG) 

```{r}

#filter assay_data to remove not mapped proteins
unmapped_proteins = read.csv("ReactomeGSAHerCAMERA_14_6_21/not_found.csv", stringsAsFactors = F)
unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
se_comparison = se_comparison[unmapped_filter, ]

# NO Normalization
# perform normalization
#normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
#assay(se_comparison) = normalized_data

# get gene sets
# protein centric copy of reactome protein gmt
gmt.file = "/home/joseph/experiments/ReactomeGMT/ReactomeGMTTest.gmt"
#gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP

# Don't filter Genes
if (F){
  # restrict gene sets and summarized experiment to interacting genes
  # restrict se and gs to intersecting genes
  igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
  if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                           "gene sets (gs) have no gene IDs in common")
  expression.data <- expression.data[igenes,]
  gs <- lapply(gs, function(s) s[s %in% igenes])
}

### USE CAMERA HERE

## verbatim code from GSA/sbea (at first)
se = se_comparison

# design matrix
grp <- colData(se)[, configEBrowser("GRP.COL")]
blk <- NULL
BLK.COL <- configEBrowser("BLK.COL")
if(BLK.COL %in% colnames(colData(se))) blk <- colData(se)[,BLK.COL]

group <- factor(grp)
paired <- !is.null(blk)
f <- "~" 
if(paired) 
{   
  block <- factor(blk)
  f <- paste0(f, "block + ") 
}   
f <- formula(paste0(f, "group"))
design <- model.matrix(f)

y <- assay(se_comparison)

# Performe Mapping

# traditional way
#gs.index <- limma::ids2indices(gs, rownames(y))

# reactome gsa method
mapping = read.csv("ReactomeGSAHerCAMERA_14_6_21/mapping.csv", stringsAsFactors = F)
identifier_mapping <- mapping$Found.identifier
names(identifier_mapping) <- mapping$Submitted.identifier
gs.index <- ids2indices_multiple(identifier_mapping, gs, y)

# run roast / camera
camera.result <- limma::camera(y, gs.index, design, sort=FALSE)
colnames(camera.result)

# formal results
gsa.result <- camera.result[, c("Direction", "FDR", "PValue", "NGenes")]
gsa.result$Pathway <- rownames(gsa.result)
gsa.result <- merge(gsa.result, gs2df(gs), by.x = "Pathway", by.y = "gene.set")

get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"], na.rm = T)
}

gs2df <- function(gs){
  df <- as.data.frame(cbind(names(gs),gs))
  colnames(df) <- c("gene.set","items")
  df$gene.set <- sapply(df$gene.set, toString)
  df
}

# Add both our pathway changes and theirs
gsa.result$afc <- unlist(lapply(gsa.result$items, function(x){get_AFC(se_comparison, x)}))

fold_changes <- rowData(se_comparison)
colnames(fold_changes)
colnames(fold_changes) <- c("Identifier", "GeneName","PVAL","ADJ.PVAL", "logFC")
gsa.result <- add_pathway_foldchanges(gsa.result, fold_changes, gs.index, assay(se_comparison))

sum(gsa.result$PValue<0.05)
sum(gsa.result$FDR<0.05)
hist(gsa.result$PValue)
```


Having successfully run CAMERA locally, we need to add our set information to the results. 

```{r}
# use annotation file to get names
annotations = read.csv("/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome_set_info.csv", stringsAsFactors = F)
colnames(annotations) = c("Pathway","Name", "description","items","count","link")
gsa.result <- merge(gsa.result, annotations, by = "Pathway")
```

We're then ready to load the CAMERA results from ReactomeGSA and plot the results

```{r}
camera_results_her = readxl::read_xlsx("ReactomeGSAHerCAMERA_14_6_21/b291df6e-cca1-11eb-a2cc-3e4fca4d787e.xlsx")

colnames(camera_results_her)
colnames(gsa.result)
head(gsa.result)
head(camera_results_her)

# proteins
both_results = merge(gsa.result, camera_results_her, by = "Name")
# genes 
# both_results = merge(gsa.result, camera_results_her, by.x = "Pathway", by.y = "Name")

plot(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$av_foldchange)
plot(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$PValue))

cor(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
cor(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
cor(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$PValue))

```
These are good results. 

It's worth also noting that while we are finding 1262 pathways, reactomeGSA is finding 1265 and only 973 of these have been matched. It's possible this has to do with the 

```{r}
plot(-log10(both_results$FDR.PROTEOMICS_INT_1), -log10(both_results$FDR))
abline(h = -log10(0.05))
abline(v = -log10(0.05))
```
This result is more or less perfect. I'm happy to deliver this to production and say we've replicated their methods. 

# Repeat using MassDynamics Workflow

Now the question is how similar would our results be had be produced them on our server. This is what that would look like. 

```{r}
library(EnrichmentAnalysisStepR)
experiment_home = "/home/joseph/experiments/her_data_gsea_test"
gmt_folder = "/home/joseph/experiments/BKG_gmt_output_human/9606"

md_results = enrichment_workflow_step(experiment_home, 
                                   gmt_folder = gmt_folder,
                                   output_folder = file.path(experiment_home,"camera"),
                                   by = "Protein",
                                   method = "camera")

md_results_reactome <- md_results$`C - P`$Reactome
```


```{r}
camera_results_her = readxl::read_xlsx("ReactomeGSAHerCAMERA_14_6_21/b291df6e-cca1-11eb-a2cc-3e4fca4d787e.xlsx")

colnames(camera_results_her)
colnames(md_results_reactome)


# proteins
both_results = merge(md_results_reactome, camera_results_her, by.x = "name", by.y = "Name")
# genes 
# both_results = merge(gsa.result, camera_results_her, by.x = "Pathway", by.y = "Name")

plot(both_results$NGenes.PROTEOMICS_INT_1, both_results$observed)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
plot(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$pval))

cor(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
cor(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
cor(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$pval))

```

```{r}
plot(-log10(both_results$FDR.PROTEOMICS_INT_1), -log10(both_results$adj.pval))
abline(h = -log10(0.05))
abline(v = -log10(0.05))
```


So it seems like what we're seeing here is diagnostic of the distinction in the isoform mapping changes. In the lower RHS quadrant we see detection made by Reactome GSA which are possible because of the multi-isoform mapping code. 

In terms of the most significant results, there seems to be agreement. As a beta feature, this is likely an acceptable oversight.

# ReactomeGSA vs MD "ReactomeGSA Hybrid" without complex mapping



```{r}
upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Protein"
method = "camera"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)

protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)
```

```{r}
write.table(assay(se_comparison), "her_data_14_6_21.tsv",  sep ="\t", quote = F, row.names = T)
# I then manually add the column name for the index
```


```{r}

#filter assay_data to remove not mapped proteins
unmapped_proteins = read.csv("ReactomeGSAHerCAMERA_14_6_21/not_found.csv", stringsAsFactors = F)
unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
se_comparison = se_comparison[unmapped_filter, ]

# NO Normalization
# perform normalization
#normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
#assay(se_comparison) = normalized_data

# get gene sets
# protein centric copy of reactome protein gmt
gmt.file = "/home/joseph/experiments/ReactomeGMT/ReactomeGMTTest.gmt"
#gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP

# Don't filter Genes
if (F){
  # restrict gene sets and summarized experiment to interacting genes
  # restrict se and gs to intersecting genes
  igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
  if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                           "gene sets (gs) have no gene IDs in common")
  expression.data <- expression.data[igenes,]
  gs <- lapply(gs, function(s) s[s %in% igenes])
}

### USE CAMERA HERE

## verbatim code from GSA/sbea (at first)
se = se_comparison

# design matrix
grp <- colData(se)[, configEBrowser("GRP.COL")]
blk <- NULL
BLK.COL <- configEBrowser("BLK.COL")
if(BLK.COL %in% colnames(colData(se))) blk <- colData(se)[,BLK.COL]

group <- factor(grp)
paired <- !is.null(blk)
f <- "~" 
if(paired) 
{   
  block <- factor(blk)
  f <- paste0(f, "block + ") 
}   
f <- formula(paste0(f, "group"))
design <- model.matrix(f)

y <- assay(se_comparison)

# Performe Mapping

# traditional way
gs.index <- limma::ids2indices(gs, rownames(y))

# reactome gsa method
#mapping = read.csv("ReactomeGSAHerCAMERA_14_6_21/mapping.csv", stringsAsFactors = F)
#identifier_mapping <- mapping$Found.identifier
#names(identifier_mapping) <- mapping$Submitted.identifier
#gs.index <- ids2indices_multiple(identifier_mapping, gs, y)

# run roast / camera
camera.result <- limma::camera(y, gs.index, design, sort=FALSE)
colnames(camera.result)

# formal results
gsa.result <- camera.result[, c("Direction", "FDR", "PValue", "NGenes")]
gsa.result$Pathway <- rownames(gsa.result)
gsa.result <- merge(gsa.result, gs2df(gs), by.x = "Pathway", by.y = "gene.set")

get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"], na.rm = T)
}

gs2df <- function(gs){
  df <- as.data.frame(cbind(names(gs),gs))
  colnames(df) <- c("gene.set","items")
  df$gene.set <- sapply(df$gene.set, toString)
  df
}

# Add both our pathway changes and theirs
gsa.result$afc <- unlist(lapply(gsa.result$items, function(x){get_AFC(se_comparison, x)}))

fold_changes <- rowData(se_comparison)
colnames(fold_changes)
colnames(fold_changes) <- c("Identifier", "GeneName","PVAL","ADJ.PVAL", "logFC")
gsa.result <- add_pathway_foldchanges(gsa.result, fold_changes, gs.index, assay(se_comparison))

sum(gsa.result$PValue<0.05)
sum(gsa.result$FDR<0.05)
hist(gsa.result$PValue)
```


Having successfully run CAMERA locally, we need to add our set information to the results. 

```{r}
# use annotation file to get names
annotations = read.csv("/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome_set_info.csv", stringsAsFactors = F)
colnames(annotations) = c("Pathway","Name", "description","items","count","link")
gsa.result <- merge(gsa.result, annotations, by = "Pathway")
```

We're then ready to load the CAMERA results from ReactomeGSA and plot the results

```{r}
camera_results_her = readxl::read_xlsx("ReactomeGSAHerCAMERA_14_6_21/b291df6e-cca1-11eb-a2cc-3e4fca4d787e.xlsx")

colnames(camera_results_her)
colnames(gsa.result)
head(gsa.result)
head(camera_results_her)

# proteins
both_results = merge(gsa.result, camera_results_her, by = "Name")
# genes 
# both_results = merge(gsa.result, camera_results_her, by.x = "Pathway", by.y = "Name")

plot(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$av_foldchange)
plot(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$PValue))

cor(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
cor(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
cor(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$PValue))

```

This results suggests that the hypothesis about multimapping causing the issue was false. 
# ReactomeGSA vs MD "ReactomeGSA Hybrid" without reactome GMT

```{r}
upload_folder = "/home/joseph/experiments/her_data_gsea_test/"
by = "Protein"
method = "camera"
perm = 1000

# Get Experiment Data
protein_viz_path = file.path(upload_folder,"protein_viz.json")
stopifnot(file.exists(protein_viz_path))
protein_viz = read_json(protein_viz_path, simplifyVector = TRUE)
stopifnot(dim(protein_viz)[1]>0)

protein_int_path = file.path(upload_folder,"proteinGroups_quant_intensities.txt")
if (!file.exists(protein_int_path)){
  protein_int_path = file.path(upload_folder,"Protein_Intensity.txt")
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
  protein_int = disco_protein_int_to_byo_int(protein_int)
} else {
  stopifnot(file.exists(protein_int_path))
  protein_int = read.csv(protein_int_path, sep ="\t")
}

protein_int <- get_protein_quant_intensities(protein_int)

# construct summarized experiment:
comparison = protein_viz[1,"conditionComparison"]
print(comparison)

# get protein viz data
comparison_viz = as.data.frame(protein_viz[1, "data"])

#Get the rest of the data

assay_data <- filter_int_by_viz(comparison_viz, protein_int)
assay_data <- filter_prot_int_cols_by_comp(comparison, assay_data)
cls_vec <- get_cls_vec(comparison, assay_data)
se_comparison <- md_to_summarized_experiment(comparison_viz, assay_data, cls_vec, by = by)
```

```{r}
write.table(assay(se_comparison), "her_data_14_6_21.tsv",  sep ="\t", quote = F, row.names = T)
# I then manually add the column name for the index
```


```{r}

#filter assay_data to remove not mapped proteins
unmapped_proteins = read.csv("ReactomeGSAHerCAMERA_14_6_21/not_found.csv", stringsAsFactors = F)
unmapped_filter = !(rownames(assay(se_comparison)) %in% unmapped_proteins$Not.found)
se_comparison = se_comparison[unmapped_filter, ]

# NO Normalization
# perform normalization
#normalized_data = limma::normalizeBetweenArrays(assay(se_comparison))
#assay(se_comparison) = normalized_data

# get gene sets
# protein centric copy of reactome protein gmt
#gmt.file = "/home/joseph/experiments/ReactomeGMT/ReactomeGMTTest.gmt"
gmt.file = "/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome.gmt"
gs <- getGenesets(gmt.file)


# create group
se_comparison$GROUP

# Don't filter Genes
if (F){
  # restrict gene sets and summarized experiment to interacting genes
  # restrict se and gs to intersecting genes
  igenes <- intersect(rownames(expression.data), unique(unlist(gs)))
  if(!length(igenes)) stop("Expression dataset (se)", " and ", 
                           "gene sets (gs) have no gene IDs in common")
  expression.data <- expression.data[igenes,]
  gs <- lapply(gs, function(s) s[s %in% igenes])
}

### USE CAMERA HERE

## verbatim code from GSA/sbea (at first)
se = se_comparison

# design matrix
grp <- colData(se)[, configEBrowser("GRP.COL")]
blk <- NULL
BLK.COL <- configEBrowser("BLK.COL")
if(BLK.COL %in% colnames(colData(se))) blk <- colData(se)[,BLK.COL]

group <- factor(grp)
paired <- !is.null(blk)
f <- "~" 
if(paired) 
{   
  block <- factor(blk)
  f <- paste0(f, "block + ") 
}   
f <- formula(paste0(f, "group"))
design <- model.matrix(f)

y <- assay(se_comparison)

# Performe Mapping

# traditional way
#gs.index <- limma::ids2indices(gs, rownames(y))

# reactome gsa method
mapping = read.csv("ReactomeGSAHerCAMERA_14_6_21/mapping.csv", stringsAsFactors = F)
identifier_mapping <- mapping$Found.identifier
names(identifier_mapping) <- mapping$Submitted.identifier
gs.index <- ids2indices_multiple(identifier_mapping, gs, y)

# run roast / camera
camera.result <- limma::camera(y, gs.index, design, sort=FALSE)
colnames(camera.result)

# formal results
gsa.result <- camera.result[, c("Direction", "FDR", "PValue", "NGenes")]
gsa.result$Pathway <- rownames(gsa.result)
gsa.result <- merge(gsa.result, gs2df(gs), by.x = "Pathway", by.y = "gene.set")

get_AFC <- function(se_comparison, gene_set){
  mean(rowData(se_comparison)[rownames(se_comparison) %in% gene_set,"FC"], na.rm = T)
}

gs2df <- function(gs){
  df <- as.data.frame(cbind(names(gs),gs))
  colnames(df) <- c("gene.set","items")
  df$gene.set <- sapply(df$gene.set, toString)
  df
}

# Add both our pathway changes and theirs
gsa.result$afc <- unlist(lapply(gsa.result$items, function(x){get_AFC(se_comparison, x)}))

fold_changes <- rowData(se_comparison)
colnames(fold_changes)
colnames(fold_changes) <- c("Identifier", "GeneName","PVAL","ADJ.PVAL", "logFC")
gsa.result <- add_pathway_foldchanges(gsa.result, fold_changes, gs.index, assay(se_comparison))

sum(gsa.result$PValue<0.05)
sum(gsa.result$FDR<0.05)
hist(gsa.result$PValue)
```


Having successfully run CAMERA locally, we need to add our set information to the results. 

```{r}
# use annotation file to get names
annotations = read.csv("/home/joseph/experiments/BKG_gmt_output_human/9606/Reactome_set_info.csv", stringsAsFactors = F)
colnames(annotations) = c("Pathway","Name", "description","items","count","link")
gsa.result <- merge(gsa.result, annotations, by = "Pathway")
```

We're then ready to load the CAMERA results from ReactomeGSA and plot the results

```{r}
camera_results_her = readxl::read_xlsx("ReactomeGSAHerCAMERA_14_6_21/b291df6e-cca1-11eb-a2cc-3e4fca4d787e.xlsx")

colnames(camera_results_her)
colnames(gsa.result)
head(gsa.result)
head(camera_results_her)

# proteins
both_results = merge(gsa.result, camera_results_her, by = "Name")
# genes 
# both_results = merge(gsa.result, camera_results_her, by.x = "Pathway", by.y = "Name")

plot(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$av_foldchange)
plot(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$PValue))

cor(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
cor(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
cor(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$PValue))

```

This results suggests that the hypothesis about multimapping causing the issue was false. 

# Test With new GMT


Now the question is how similar would our results be had be produced them on our server. This is what that would look like. 

```{r}
library(EnrichmentAnalysisStepR)
experiment_home = "/home/joseph/experiments/her_data_gsea_test"
gmt_folder = "./GMTs/MD/"

md_results = enrichment_workflow_step(experiment_home, 
                                   gmt_folder = gmt_folder,
                                   output_folder = file.path(experiment_home,"camera"),
                                   by = "Protein",
                                   method = "camera")

md_results_reactome <- md_results$`C - P`$Reactome
```


```{r}
camera_results_her = readxl::read_xlsx("ReactomeGSAHerCAMERA_14_6_21/b291df6e-cca1-11eb-a2cc-3e4fca4d787e.xlsx")

colnames(camera_results_her)
colnames(md_results_reactome)


# proteins
both_results = merge(md_results_reactome, camera_results_her, by.x = "name", by.y = "Name")
# genes 
# both_results = merge(gsa.result, camera_results_her, by.x = "Pathway", by.y = "Name")

plot(both_results$NGenes.PROTEOMICS_INT_1, both_results$observed)
plot(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
plot(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$pval))

cor(both_results$NGenes.PROTEOMICS_INT_1, both_results$NGenes)
cor(-both_results$av_foldchange.PROTEOMICS_INT_1, both_results$afc)
cor(-log10(both_results$PValue.PROTEOMICS_INT_1), -log10(both_results$pval))

```

Awesome!

```{r}
plot(-log10(both_results$FDR.PROTEOMICS_INT_1), -log10(both_results$adj.pval))
abline(h = -log10(0.05))
abline(v = -log10(0.05))
```

